---
title: "bulk_RNAseq_Analysis"
author: "DTG"
date:  "`r Sys.Date()`"
output: 
  rmdformats::robobook:
    fig_width: 9
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE, echo=FALSE, fig.width = 8, fig.height = 4)
library(tidyverse)
library(ggforce)
library(grid)
library(ggpubr)
library(yaml)
library(DESeq2)
library(EDASeq)
library(DT)
library(shiny)
library(EnhancedVolcano)
library(tools)
library(ComplexHeatmap)
library(gridExtra)
library(gtable)
library(circlize)
library(openxlsx)
library(reshape2)
library(kableExtra)
library(rlang)
library(forcats)
library(gtools)
library(RColorBrewer)
library(fgsea)
library(MatrixGenerics)
#library(plyr)

counts <- DESeq2::counts
#config <- yaml.load_file("config.yaml")
source("functions.R")
```

<style type="text/css">
.book .book-body .page-inner {
  max-width: 1600px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r}
analysis <- readRDS('analysis_post_QC.Rds')
analysis$config <- yaml.load_file("config.yaml")


## need to make within group individual tags to account for individual differences
## doesn't work to just pass the individual tag in the smple table (linear combination of factors)
# for (group in analysis$sampleTable$Group){
#   members <- unique(analysis$sampleTable %>% filter(Group == group) %>% select(Individual))
#   indices <- analysis$sampleTable$Group == group
#   analysis$sampleTable$GroupID[indices] <- plyr::mapvalues(
#     analysis$sampleTable$Individual[indices],
#     from = members$Individual,
#     to = 1:nrow(members)
#   )
# }
analysis$sampleTable <- data.frame(lapply(analysis$sampleTable, as.factor))
```

## QC {.tabset}

### PCA

```{r, fig.height=6, fig.width=8}
print(analysis$pca)
```

### Sample Info

```{r}
knitr::kable((analysis$sampleTable), position='center', table.attr="style='width:100%;'", format='html') %>%
    kable_styling(full_width = TRUE)
```

```{r}
## one last safety check to make sure metadata and samples are properly associated
stopifnot(colnames(analysis$ddsDrop) == analysis$sampleTable$SampleID)

analysis <- readRDS('analysis_post_QC.Rds')
analysis$ddsDrop <- DESeqDataSetFromMatrix(counts(analysis$ddsDrop),
                                           analysis$sampleTable,
                                           design = as.formula(paste0("~ ", analysis$config$design)))
analysis$ddsDrop <- DESeq(analysis$ddsDrop)
                                           
```

## DGE {.tabset}

This section of the report analyses gene expression between groups of samples within the study. Changes in gene expression between groups are usually reported on a log2 scale. The significance of a detected change in expression is reported in a raw/nominal p-value and a multiple-testing corrected/adjusted p-value that accounts for the large number of tests conducted in differential gene expression (DGE) and the inherent false positive rate assumed with frequentest p-values.

Each tab has results from a specific comparison within the study. Each tab shows 

1. A summary of significantly differentially expressed genes (DGE)

2. A volcano plot of DGE

3. A table of all significant DGE

4. A dotplot of top GSEA enriched biological pathways

5. A table of top GSEA enriched biological pathways

The order of terms in tab labeling informs how to interpret results. For log-fold change and enrichment, the first listed group is the numerator and the second is the denominator. So in a comparison of X vs Y, a positive log-fold change means the gene is more highly expressed in X, while a negative change means the gene is more highly expressed in Y. enrichment is similar: positive enrichment relates to enrichment in X, and negative enrichment indicates greater enrichment in Y. 

Gene Set Enrichment Analysis (GSEA) is a statistical approach to test systematic changes in groups of genes corresponding to biological pathways. A pathway is generally 'enriched' in one study group if the genes in that pathway are systematically up or down regulated in a manner and magnitude that is unlikely to be due to chance. Read more about GSEA [here](https://www.gsea-msigdb.org/gsea/index.jsp)

***

### Overview

```{r, fig.height = 4, fig.width=8}
# timecourse_comparison_overview <- function(time, alpha=0.05){
#   
#   a <- results(analysis$ddsDrop, alpha=alpha, contrast=list(c(paste0('Time_D', time, '_vs_D0'), paste0('GroupsamRNA.TimeD', time))), test = 'Wald', filter = maxMinFilter(analysis$ddsDrop, intgroup = 'PCA.elipse', comp = c(paste0('samRNA_D', time), 'samRNA_D0')))
#   b <- nrow(as.data.frame(a) %>% filter(padj<alpha))
#   a <- results(analysis$ddsDrop, alpha = alpha, name=paste0("Time_D",time,"_vs_D0"), test="Wald",
#                                          filter = maxMinFilter(analysis$ddsDrop,
#                                          intgroup = 'PCA.elipse',
#                                          comp = c(paste0('LNP-RNA_D',time),"LNP-RNA_D0"))) 
#   d <- nrow(as.data.frame(a) %>% filter(padj<alpha))# %>% filter(log2FoldChange<0))
#   return(list(samRNA=b, LNP.RNA=d))
# }
# 
# res <- lapply(list('1','2','4','7'), timecourse_comparison_overview)
# res <- as.data.frame(lapply(res, unlist), col.names=paste0('D', c(1,2,4,7)))
# #rownames(res) <- c('samRNA', 'LNP-RNA')
# res <- melt(res %>% rownames_to_column('DIR'), id.vars='DIR')
# res$variable <- as.numeric(gsub('D', '', res$variable))
# ggplot(res, aes(x=variable,y=value, group=DIR, color=DIR)) + geom_line() + labs(x="Days after intervention", y="# of significantly diff. expressed genes", color="Treatment group", title="DGE at each day compared to pre-infection", caption="DGE is calculated on comparing individuals within a treatment to the pre-treatment state (D0)") + theme_minimal() + theme(legend.position=c(0.9,.8)) + scale_x_continuous(breaks=c(1,2,4,7))
```


```{r, fig.height = 4, fig.width=8}
# timecourse_comparison_overview <- function(time, alpha=0.05){
#   a <- results(analysis$ddsDrop, alpha=alpha, name=paste0("GroupsamRNA.TimeD",time), test="Wald", filter = maxMinFilter(analysis$ddsDrop, intgroup = 'PCA.elipse', comp = c(paste0('samRNA_D',time), 'samRNA_D0')))
#   b <- nrow(as.data.frame(a) %>% filter(padj<alpha))
#   return(b)
# }
# 
# res <- lapply(list('1','2','4','7'), timecourse_comparison_overview)
# res <- as.data.frame(lapply(res, unlist), col.names=paste0('D', c(1,2,4,7)))
# #rownames(res) <- c('samRNA', 'LNP-RNA')
# res <- melt(res %>% rownames_to_column('DIR'), id.vars='DIR')
# res$variable <- as.numeric(gsub('D', '', res$variable))
# ggplot(res, aes(x=variable,y=value)) + geom_line() + labs(x="Days after intervention", y="# of significantly diff. expressed genes", title="DGE comparing LNP-RNA to samRNA groups at each collection date", caption="This shows significant interaction effects of time and treatment at each timepoint") + theme_minimal() + theme(legend.position=c(0.9,.8)) + scale_x_continuous(breaks=c(1,2,4,7))
```

###

```{r GSEA_setup}
## read GMT files
gmt.file <- list()
  for (a in file.path(analysis$config$rootDir, analysis$config$gseaFilesPath, analysis$config$gseaCollections)){#'mh.all.v0.3.symbols.gmt'
    gmt.file <- append(gmt.file, gmtPathways(file.path(a)))
  }
```

```{r run_comparisons}

### Add comparisons as you would pass them to a DESeq2::results contrast perameter E.g., 
### if you want a single existing contrast, pass it in as a string
### if you want to make a new contrast between factors, make it a list within the list

# resultsNames(analysis$ddsDrop)

  
my_comparisons <- list('Group_FAC_vs_Cont', 'Group_LiproxposFAC_vs_Cont', 'Group_MTposFAC_vs_Cont')

## groups to include for greg's filter function, specific to each comparison
filter_on_list <- list(c('FAC', 'Cont'),
                       c('LiproxposFAC', 'Cont'),
                       c('MTposFAC', 'Cont'))
                       
pairwise_comparisons <- function(x, filt_groups, intgroup, alpha=0.05){
  ## Generate DESeq2 results objects for contrasts of interest
  if (class(x) == 'list'){
    result <- results(analysis$ddsDrop, contrast=x, test='Wald', alpha=alpha, filter = maxMinFilter(analysis$ddsDrop, intgroup = intgroup, comp = filt_groups))
  } else if (class(x) == 'character'){
    result <- results(analysis$ddsDrop, name=x, test='Wald', alpha=alpha, filter = maxMinFilter(analysis$ddsDrop, intgroup = intgroup, comp = filt_groups))
  }
  result@metadata$contrast <- paste0(unlist(x), collapse='_')
  return(result)
}

## Get DESeq2 results for each comparison of interest
## intgroup specifies which metadata column your filter_on_list refers to
results <- mapply(pairwise_comparisons, my_comparisons, filter_on_list, intgroup = 'Group')

## Convert these result objects to custom class with visualizations, etc
results <- lapply(results, my_results_transform, wb)

## Name result objects from contrasts
names(results) <- lapply(results, function(x) x@metadata$contrast)

```

```{r prepare_results}
### you can add custom ggplot tracks to the outputs here, or edit datatables
### Input is the modifided DESeq2 object with the visualizations
output_results <- function(comp){
  # Just plaintext summary of the result object
  summary(results[[comp]])
  # Fixed window volcano plot
  volcanoPlot <- results[[comp]]@visualizations$volplot_fixed + labs(caption='Cutoffs for significance\np-value of 0.05\n1.3 fold-change')
  # Interactive datatable of all DGE results for this comp
  dge_data <- data.frame(results[[comp]]) %>% select(baseMean, log2FoldChange, lfcSE, padj) %>% filter(padj < analysis$config$alpha) %>% mutate(baseMean=round(baseMean), log2FoldChange = round(log2FoldChange,2), lfcSE = round(lfcSE, 2), padj=round(padj,4)) %>% na.omit() %>% rownames_to_column()
  DGEtable <- DT::datatable(dge_data, rownames = FALSE, colnames = c('Gene', 'Base mean expression', 'Log2 fold change', 'Fold change std. error', 'Adj. p-value'), caption='DESeq2 output for this comparison, significant genes only', filter='top', autoHideNavigation = TRUE,extensions = c('Buttons', 'Scroller'),  options = list(
    dom = 'Bfrtip',
    buttons = c('copy', 'csv'),
    deferRender = TRUE
  )) 
  # Dotplot of top hallmark GSEA pathways
  gseaDotplot <- gsea_dotplot(results[[comp]]@gsea)
  
  # Interactive table of GSEA results
  gsea_data <- results[[comp]]@gsea %>% filter(size>5) %>% arrange(pval) %>% group_by(source) %>% slice_head(n=25) %>% select(pathway, source, pval,padj,NES,size) %>% arrange(pval)
  gseaTable <- DT::datatable(gsea_data, rownames = FALSE, filter='top', autoHideNavigation = TRUE, caption='GSEA enrichment results, top from each source') %>% formatRound(c('pval','padj','NES'), digits=3)
  
  geneList <- rownames(head(data.frame(results[[comp]]) %>% filter(!is.na(pvalue)) %>% filter(padj<analysis$config$alpha) %>% filter(abs(log2FoldChange)>log2(1.5)) %>% arrange(pvalue), 30))
  
  topGenes <- heatmap_from_genelist(geneList, 'Group', 'Cont', data=analysis$rldDrop)
  
  return(list(volcanoPlot=volcanoPlot,DGEtable=DGEtable,gseaDotplot=gseaDotplot,gseaTable=gseaTable, topGenes= topGenes))
}
```

### Comp1

`r paste0(unlist(my_comparisons[[1]]), collapse="_")`

```{r}
### call results outputs here with output_results(name_of_comparison)
### comparison names are based on the input lists earlier
comparison <- output_results(paste0(unlist(my_comparisons[[1]]), collapse="_"))
comparison$volcanoPlot
comparison$DGEtable
comparison$topGenes
comparison$gseaDotplot
comparison$gseaTable

```


### Comp2
<!-- You can make this the active tab instead of the first one if you want to highlight certain comp -->

`r paste0(unlist(my_comparisons[[2]]), collapse="_")`

```{r}
### call results outputs here with output_results(name_of_comparison)
### comparison names are based on the input lists earlier
comparison <- output_results(paste0(unlist(my_comparisons[[2]]), collapse="_"))
comparison$volcanoPlot
comparison$DGEtable
comparison$topGenes
comparison$gseaDotplot
comparison$gseaTable

```

***

## Genes of interest {.tabset}

These heatmaps show the relative expression (regularized and log2 transformed) of genes for each sample normalized to the baseline (mean expression for pre-treatment samples). Individual samples are grouped within their treatment/timepoint. If you have genes you are interested in seeing from your study, we can plot them here.

### Top genes from whole study

```{r, fig.width = 10, fig.height=5}
## creating an arbitrary genelist
top_genes <- data.frame(results(analysis$ddsDrop)) %>% filter(!is.na(pvalue)) %>% filter(abs(log2FoldChange)>log2(1.5))
top_genes <- top_genes[order(top_genes$pvalue),]
geneList <- (rownames(head(top_genes, 20)))

## Pass in a genelist
heatmap_from_genelist(geneList, baseline_grouping = 'Group', baseline = 'Cont')
```

***

## GSEA

Top GSEA pathways from this study. If you have specific biological pathways you want to see, you can request them to be plotted here.

```{r, fig.width=7, fig.height=9}
tests <- data.frame(x = c(
  'Group_FAC_vs_Cont',
  'Group_LiproxposFAC_vs_Cont',
  'Group_MTposFAC_vs_Cont'
), name = c(
  'FAC_vs_Cont',
  'LiproxposFAC_vs_Cont',
  'MTposFAC_vs_Cont'
))

## number to pull from each comparison
n <- 5
bytmp <- lapply(tests$x, function(x) {
  a <- results[[x]]@gsea
  #a <- a %>% mutate(test=((-log10(pval))*abs(NES)))
  head(a %>% arrange(pval), n)$pathway
} )
tmp <- lapply(tests$x, function(x) {
  a <- results[[x]]@gsea
  #a <- a %>% mutate(test=((-log10(pval))*abs(NES)))
  a %>% filter(pathway %in% unlist(bytmp))
} )
tmp2 <- data.table::rbindlist(tmp, idcol=TRUE)

tmp2$`.id` <- plyr::mapvalues(tmp2$`.id`, from= 1:nrow(tests), to = tests$name)
tmp2$pathway <- wrap_underscore_strings_balance(tmp2$pathway,36)

  ggplot(tmp2) + geom_point(aes(x=`.id`, y=pathway, size=-log(pval), color=NES)) +
    scale_color_gradient2(low="blue", mid="white",high="red", midpoint=0, breaks=c(-2,-1,0,1,2), limits=c(min(tmp2$NES,-1), max(tmp2$NES,1))) +
    theme_classic(11) +
    theme(panel.grid.major = element_line(colour = "grey92"),
          panel.grid.minor = element_line(colour = "grey92"),
          panel.grid.major.y = element_line(colour = "grey92"),#element_blank(),
          panel.grid.minor.y = element_line(colour = "grey92")) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    labs(x="Comparison", y="Gene set", color = "Normalized\nenrichment\nscore", size="Nom p-val", title="Top Pathways") +  scale_radius(name="NOM p-val", range=c(1,8), breaks=-log10(c(0.1,0.01,0.001,0.0001)), labels=c(0.1,0.01,0.001,0.0001)) 

```

```{r}
results_of_interest <- data.frame(comp = c(
  'Group_FAC_vs_Cont',
  'Group_LiproxposFAC_vs_Cont',
  'Group_MTposFAC_vs_Cont'
), label = c(
  'FAC_vs_Cont',
  'LiproxposFAC_vs_Cont',
  'MTposFAC_vs_Cont'
))

master_table <- data.frame(baseMean = results[[1]]$baseMean, row.names = rownames(results[[1]]))
for (i in 1:nrow(results_of_interest)){
  comp <- results_of_interest$comp[i]
  label <- results_of_interest$label[i]
  res <- as.data.frame(results[[comp]]) %>% select(c(log2FoldChange,lfcSE, pvalue, padj)) %>% rename(log2FoldChange = 'Log2 fold change' , lfcSE = 'Log fold change standard error', pvalue = 'P value Wald test', padj = 'Benjamini-Hochberg corrected p value') %>% rename_with(~ paste0(label,':\n', .x))
  master_table <- cbind(master_table, res)
}
master_table <- master_table %>% rename(baseMean = 'mean of normalized counts' ) %>% rownames_to_column('Gene')

write_excel_csv(master_table, file.path('outputs', paste0(analysis$config$analysis, '_DESeq2_results.csv')))
```

```{r}
results_of_interest <- data.frame(comp = c(
  'Group_FAC_vs_Cont',
  'Group_LiproxposFAC_vs_Cont',
  'Group_MTposFAC_vs_Cont'
), label = c(
  'FAC_vs_Cont',
  'LiproxposFAC_vs_Cont',
  'MTposFAC_vs_Cont'
))
master_table <- as.data.frame(t(as.data.frame(lapply(gmt.file, length))))
colnames(master_table) <- 'size'
master_table <- master_table %>% rownames_to_column('pathway')
#master_table <- data.frame(size = results[[1]]@gsea$size, pathway = results[[1]]@gsea$pathway)
for (i in 1:nrow(results_of_interest)){
  comp <- results_of_interest$comp[i]
  label <- results_of_interest$label[i]
  res <- results[[comp]]@gsea %>% select(pathway,pval, padj, NES) %>% column_to_rownames('pathway') %>% rename(pval = 'p-value', padj = 'Benjamin-Hochberg adjusted p-value', NES='Normalized enrichment score') %>% rename_with( ~ paste0(label, ':\n', .x)) %>% rownames_to_column('pathway')
  master_table <- merge(master_table, res, by='pathway', all=TRUE)
}

master_table$source <- unlist(lapply(master_table$pathway, function(x){str_split(x, '_')[[1]][1]}))

master_table <- master_table %>% relocate(pathway, source, size)
write_excel_csv(master_table, file.path('outputs', paste0(analysis$config$analysis, '_GSEA_results.csv')))
```




<!-- ```{r} -->
<!-- A.fix <- as.data.frame(analysis$assayRlogForGSEA) %>% rownames_to_column('Name') %>% filter(!duplicated(Name)) %>% column_to_rownames('Name') %>% data.matrix()  -->
<!-- gene.labels <- rownames(A.fix) -->
<!-- sample.names <- colnames(A.fix) -->

<!-- phen_map <- as.factor(plyr::mapvalues(sample.names, from=analysis$ddsDrop@colData$SampleID, to=as.character(analysis$ddsDrop@colData$Group))) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- comp.num <-  'FAC' -->
<!-- comp.denom <- 'LiproxposFAC' -->
<!-- #sample.names[sample.phen=='FAC'] -->
<!-- comp.ind <- c(grep(paste0('^',comp.num,'$'), sample.phen), grep(paste0('^',comp.denom,'$'), sample.phen))#sample.phen %in% comparison -->



<!-- O <- GSEA.GeneRanking(A[,comp.ind], as.integer(phen_map[comp.ind]), gene.labels, nperm = 1, permutation.type = 0,  -->
<!--     sigma.correction = "GeneCluster", fraction = 1, replace = FALSE,  -->
<!--     reverse.sign = FALSE, rank.metric = "S2N") -->
<!-- # order.matrix[, n.starts[nk]:n.ends[nk]] <- O$order.matrix -->
<!-- # obs.order.matrix[, n.starts[nk]:n.ends[nk]] <- O$obs.order.matrix -->
<!-- # correl.matrix[, n.starts[nk]:n.ends[nk]] <- O$rnk.matrix -->
<!-- # obs.correl.matrix[, n.starts[nk]:n.ends[nk]] <- O$obs.rnk.matrix -->
<!-- ranks <- O$obs.rnk.matrix[O$obs.order.matrix,] -->
<!-- ``` -->

<!-- ```{r} -->
<!-- res <- fgseaSimple(pathways=gmt.file, -->
<!--                    stats=ranks, -->
<!--                    nperm=1000) -->
<!-- ``` -->

