---
title: "bulk_RNA-seq_analysis"
author: "DTG"
date: '`r format(file.info("p22074_Peijian_Analysis.Rmd")$ctime, "%Y-%m-%d")`'
output: 
  html_notebook: 
    code_folding: hide
---

```{r setup}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(tidyverse)
library(ggforce)
library(grid)
library(ggpubr)
library(yaml)
library(DESeq2)
library(EDASeq)
library(DT)
library(shiny)
library(EnhancedVolcano)
library(tools)
library(ComplexHeatmap)
library(gridExtra)
library(gtable)
library(circlize)
library(openxlsx)
library(reshape2)
library(kableExtra)
library(rlang)
#library(plyr)
source("/yerkes-cifs/runs/home/gktharp/gitrepos/Coding_Club/BulkRNASeq/gktharp/GKTDESeq2Extentions.R")
counts <- DESeq2::counts
analysis = list()
```

```{r reference_files}
config <- yaml.load_file("config.yaml")
analysis$config <- config
exp.design <- read.table("exp_design_template.txt", header=T, stringsAsFactors = T, colClasses="factor")
exp.design <- exp.design %>% dplyr::arrange(across(all_of(analysis$config$sampleGrouping)))
knitr::kable((exp.design), position='center', table.attr="style='width:100%;'", format='html') %>%
    kable_styling(full_width = T)
```


```{r samples}
analysis$samplefileIDs = dir(config$alignmentDir)
analysis$sampleSTARReads <- sapply(analysis$samplefileIDs, function(sid) {read_tsv(paste0(
  dir(config$alignmentDir, pattern = sid, full.names = TRUE),
  "/",sid,
  config$STARreadSuffix
), col_names = c("gene_id","unstranded_count","sense_count","antisense_count"))},
simplify = FALSE,
USE.NAMES = TRUE)
analysis$sampleTable <- exp.design


```

```{r dds}

analysis$mapBins <- do.call(cbind,
                                   sapply(analysis$samplefileIDs,                                      function(sid){analysis$sampleSTARReads[[sid]][,"unstranded_count"][c(1:4),]}))
rownames(analysis$mapBins) <- analysis$sampleSTARReads[[1]][c(1:4),][["gene_id"]]
colnames(analysis$mapBins) <- recode(as.character(unlist(strsplit(analysis$samplefileIDs, ".unstranded_count"))), !!!setNames(exp.design$SampleID, as.character(exp.design$FileID)))
 
raw_counts <- do.call(cbind,sapply(exp.design$FileID, function(sid){analysis$sampleSTARReads[[sid]][,"unstranded_count"][-c(1:4),]}))
colnames(raw_counts) <- exp.design$SampleID
analysis$dds <- DESeqDataSetFromMatrix(raw_counts,
  analysis$sampleTable,
  design = ~ 1
)

raw_counts <- DESeqDataSetFromMatrix(countData = DESeq2::counts(analysis$dds[,!analysis$dds@colData$SampleID %in% analysis$config$dropSamples]), colData = analysis$dds@colData, design = as.formula(paste0("~ ", analysis$config$design)))


```


# Mapping

Mapping outcomes of reads from each sample. Reads need to be 'identified' or successfully mapped to a single feature to be used for gene expression analysis.

```{r mapping, fig.height=8}

rownames_to_column(as_tibble(analysis$mapBins, rownames = NA), var = "map_result") %>%
  pivot_longer(!map_result, names_to = "SampleID", values_to = "count") %>%
  #mutate(SampleID = str_sub(SampleID,end = -18)) %>%
  ggplot(aes(x = SampleID, y = count, fill = factor(map_result, levels = c("N_unmapped","N_multimapping","N_noFeature","N_ambiguous","N_identified")))) +
  geom_bar(stat = "identity") +
  theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) +
  guides(fill=guide_legend(title="Map Result")) +
  ggtitle(paste0(analysis$config$analysis," Mapping"))
  
  

```
```{r}



geneInfoTab <- read_tsv(file.path("/yerkes-cifs/runs/Genome_references/", analysis$config$referenceDir, "geneInfo.tab"), skip = 1, col_names = c("gene_id","gene_symbol","gene_type"))

lowcounts <- colSums(assay(raw_counts)) < nrow(geneInfoTab)
if (sum(lowcounts)>0){
  cat("Removing sample(s): ", names(lowcounts)[lowcounts], "due to low counts", sep="\n")
}
analysis$ddsDrop <- raw_counts[,!lowcounts]



rownames(analysis$ddsDrop) <- make.names(geneInfoTab$gene_symbol, unique=T)
# rownames(analysis$dds) <- analysis$sampleSTARReads[[1]][-c(1:4),][["gene_id"]]
analysis$mapBins <- rbind(analysis$mapBins,"N_identified" = colSums2(DESeq2::counts(analysis$dds)))

#analysis$dds <- DESeq(analysis$dds, parallel = TRUE)
analysis$vst <- varianceStabilizingTransformation(analysis$ddsDrop,
                                                         blind = FALSE,
                                                         fitType = "parametric")


analysis$ddsDrop <- DESeq(analysis$ddsDrop, parallel = TRUE)
```

## Sample correlation 

Calculated using normalized transcript counts for all genes

```{r sample_cor_heatmap, fig.width=8, fig.height=6}
cormat <- cor(DESeq2::counts(analysis$ddsDrop, normalized=T))
cormat <- melt(cormat)
ggplot(cormat, aes(x=Var1, y=Var2, fill=value)) + geom_tile() + scale_fill_gradient(high="red", low="white")+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ theme(axis.title = element_blank()) + labs(fill='Corr')

```

# Relative Log Expression (RLE) normalization

```{r rle, fig.height=8, fig.width=8}
# par(mar=c(18,3,2,0))
# plotRLE(counts(analysis$dds)[rowMins(counts(analysis$dds))>0,], outline = FALSE, main = "RLE Raw counts",las = 2)
# plotRLE(counts(analysis$dds[rowMins(counts(analysis$dds))>0,], normalized = TRUE), outline = FALSE, main = "RLE Norm counts", las = 2)

#rawMedians <- if_else(rowMins(counts(analysis$dds))>0,rowMedians(counts(analysis$dds)),NA_real_)
rawLogCounts <- log(DESeq2::counts(analysis$ddsDrop)[rowMins(DESeq2::counts(analysis$ddsDrop))>0,])
rawMedianLogs <- rowMedians(rawLogCounts)
rawLogRatios <- rawLogCounts - rawMedianLogs
#rawMedians <- rowMedians(counts(analysis$dds)[rowMins(counts(analysis$dds))>0,])
#rawLogRatios <- log(counts(analysis$dds)[rowMins(counts(analysis$dds))>0,]) - log(rawMedians)

normLogCounts <- log(DESeq2::counts(analysis$ddsDrop, normalized = TRUE)[rowMins(DESeq2::counts(analysis$ddsDrop))>0,])
normMedianLogs <- rowMedians(normLogCounts)
normLogRatios <- normLogCounts - normMedianLogs
#normMedians <- rowMedians(counts(analysis$dds, normalized = TRUE)[rowMins(counts(analysis$dds))>0,])
#normLogRatios <- log(counts(analysis$dds, normalized = TRUE)[rowMins(counts(analysis$dds))>0,]) - log(normMedians)

as_tibble(rawLogRatios, rownames = NA) %>% pivot_longer(everything(), names_to = "Sample", values_to = "RLE") %>% 
  ggplot(aes(x=Sample,y=RLE)) +
  geom_hline(yintercept = 0, color = "red") + geom_violin(draw_quantiles = c(0.25,0.75), trim = TRUE, color = "lightgreen", alpha = 0.1) +
  geom_boxplot(alpha = 0) + theme_bw() + theme(axis.text.x = element_text(vjust = 0.5,angle = 90),axis.title.x = element_blank(),aspect.ratio = 0.55) + ggtitle("RLE Raw") #+ scale_y_continuous(limits = c(-9,3.25), expand = c(0,0))

as_tibble(normLogRatios, rownames = NA) %>% pivot_longer(everything(), names_to = "Sample", values_to = "RLE") %>% 
  ggplot(aes(x=Sample,y=RLE)) +
  geom_hline(yintercept = 0, color = "red") + geom_violin(draw_quantiles = c(0.25,0.75), trim = TRUE, color = "lightgreen", alpha = 0.1) + 
  geom_boxplot(alpha = 0) + 
  theme_bw() + theme(axis.text.x = element_text(vjust = 0.5,angle = 90),axis.title.x = element_blank(),aspect.ratio = 0.55) + ggtitle("RLE Normalized") #+ scale_y_continuous(limits = c(-7,5.9), expand = c(0,0)) 

```


# PCA

```{r pca}
#for (i in 1:(length(colnames(analysis$sampleTable))-2)){
  #overlay <- colnames(analysis$sampleTable)[3:length(colnames(analysis$sampleTable))][i]
  x <- colnames(analysis$sampleTable)[1:(length(colnames(analysis$sampleTable))-2)]
  analysis$pca <- pcaPlotGKT(analysis$vst,
                                  intgroup = names(colData(analysis$ddsDrop)),
                                  xpc = 1, ypc = 2) +
#  geom_path(aes(group = SubjectID)) 
  geom_point(aes(color = (if (is.null(analysis$config$pcaMapping$color)) NULL else .data[[analysis$config$pcaMapping$color]]),
                 shape = (if (is.null(analysis$config$pcaMapping$shape)) NULL else .data[[analysis$config$pcaMapping$shape]])),
             size = 4) +
    labs(color=analysis$config$pcaMapping$color, shape=analysis$config$pcaMapping$shape) +
  geom_text_repel(aes(label = paste0(Individual)),size = 3, hjust = 0.5, vjust = -0.5) +
  scale_x_continuous(expand = c(0.5,0)) +
  theme_bw() + ggtitle(paste0(analysis$config$analysis," PCA"))
  print(analysis$pca)
#}

# 
# if (!is.null(analysis$config$dropSamples[[1]])) {
#   print("PCA after dropping samples")
#   analysis$pca <- pcaPlotGKT(analysis$vst[,!analysis$vst$gcid %in% analysis$config$dropSamples],
#                                   intgroup = names(colData(analysis$ddsDrop)),
#                                   xpc = 1, ypc = 2) +
# #  geom_path(aes(group = SubjectID)) +
#   geom_point(aes(color = Group), size = 4) +
#   geom_text_repel(aes(label = paste0(SampleID)),size = 3, hjust = 0.5, vjust = -1) +
# #  geom_path(aes(group = SubjectID)) +
#   scale_x_continuous(expand = c(0.5,0)) +
#   theme_bw() + ggtitle(paste0(analysis$config$analysis," PCA Drop outliers"))
# 
#   analysis$pca
# }


```

```{r write_tables}

write_csv(analysis$sampleTable, path = paste0(analysis$config$analysis,".csv"), col_names = TRUE)
write_csv(as.data.frame(DESeq2::counts(raw_counts))%>%rownames_to_column(var = "gene_id"),path = paste0("raw_count_",analysis$config$reference, "_", analysis$config$analysis,".csv"),col_names = TRUE)

## request for raw counts with EnsembleIDs 
## cant just pull the count table that has been labeled with gene symbols
## because symbols that start with a number have been prepended with "X"
## and non-unique symbols have been appended with a ".#" index
## reconstruct ensemble count table
ensembleRawCount <- do.call(cbind,sapply(analysis$samplefileIDs, function(sid){analysis$sampleSTARReads[[sid]][,"unstranded_count"][-c(1:4),]}))
rownames(ensembleRawCount) <- analysis$sampleSTARReads[[1]][-c(1:4),][["gene_id"]]
#str(ensembleRawCount)
tmp <- write_csv(as.data.frame(ensembleRawCount)%>%rownames_to_column(var = "EnsemblID"),
          path = paste0("raw_count_EnsemblID_", analysis$config$reference, analysis$config$analysis, ".csv"),col_names = TRUE)


```


# Pairwise group comparison results


```{r, fig.height=8.5, fig.width=11}

#pairs <- utils::combn(unique(analysis$sampleTable$Group),2)
#analysis$config$groupPairs
wb <- createWorkbook()
sig.genes <- list()
layout <- rbind(c(1,3,3),
                c(NA,3,3),
                c(2,2,2),
                c(2,2,2),
                c(2,2,2))
#sig.genes <- vector(mode='character')
pairwise_comparisons <- function(pair){
  f0 <- unlist(str_split(pair, "_"))
  numer_contrast <- f0[1]
  denom_contrast <- f0[3]
  result <- results(analysis$ddsDrop, contrast=c('Group', numer_contrast, denom_contrast), filter=maxMinFilter(analysis$ddsDrop, "Group", c(numer_contrast,denom_contrast)))
  volData <- result[!is.na(result$padj),]
  volData <- volData[order(volData$padj),]
  summary_out <- capture.output(summary(result))
  labs <- if (!is.null(analysis$config$geneList[[1]])) analysis$config$geneList else rownames(volData[1:20,])
  volplot1 <- (EnhancedVolcano(volData,
                  x = 'log2FoldChange',
                  y = 'padj',
                  lab = rownames(volData),
                  selectLab=labs,
                  drawConnectors = TRUE,
                  colConnectors = "lightgrey",
                  pCutoff = analysis$config$alpha,
                  FCcutoff = log2(1.3),
                  title =NULL,# paste0(numer_contrast, " vs ", denom_contrast),
                  caption=NULL,
                  subtitle=NULL
  ) )
  volplot2 <- volplot1 + ylim(c(0,100)) + xlim(c(-5,5))
  if (nchar(pair)>31){
    pair <- substr(pair,1,31)
  }
  b <- addWorksheet(wb, pair)
  #class(a)
  #print(summary_out)
  #par(mfcol=c(1,2))
  #print(d)
  #print(e)
  gVolData <- as.mutate.data.frame(result)
  gVolPlot <- gVolData %>%
  dplyr::arrange(!is.na(.data[[colnames(gVolData)[5]]]),desc(.data[[colnames(gVolData)[5]]])) %>%
  ggplot(aes(x=.data[[colnames(gVolData)[2]]],
             y=.data[[colnames(gVolData)[3]]])) +
  geom_vline(aes(xintercept = filterThreshold), color = "goldenrod") +
  geom_text(aes(x = filterThreshold, y=-5,
                label = format(filterThreshold, digits=3)),
            color = "goldenrod", hjust = 0.5, vjust = 1) +
  guides(color = guide_legend(override.aes = list(size = 2))) +
  geom_point(aes(color = significant), size = 0.1) +
  scale_x_log10() +
  theme_bw() +
  scale_color_manual(values = c("significant" = "red",
                                "not_significant" = "green",
                                "removed" = "blue",
                                "filterThreshold" = "goldenrod"), drop = FALSE)


  b <- writeData(wb, sheet=pair, x=as.data.frame(volData), rowNames=T)
  #write.xlsx(a, paste0(analysis$config$analysis, "_DESeq2_results.xlsx"), sheetName=pair)
  sig.data <- volData[volData$padj<analysis$config$alpha,]
  #summary_out <- textGrob(summary_out)
  grid.arrange(textGrob(paste(summary_out[1:8], collapse="\n")), ggarrange(volplot1, volplot2,common.legend = T), gVolPlot, ncol=2, nrow=2, layout_matrix=layout, top=textGrob(pair, gp=gpar(fontsize=25)))#, bottom="padj < 0.1, | FoldChange | > 1.3")#, fig.lab=unlist(out[[i]]$pair))
  #print(out[[i]]$vol1)
  #print(out[[i]]$vol2)
  sig.genes <- c(sig.genes, sig.data@rownames)
  return(list(sig.data@rownames, result, pair))#, vol1=d, vol2=e, summary_out=summary_out, pair=pair))
}
out <- lapply(analysis$config$groupPairs, pairwise_comparisons)
sig.genes <- unlist(lapply(out, "[", 1))
results_objs <- lapply(out, "[[", 2)
results_objs <- lapply(results_objs, as.data.frame)
pair_order <- unlist(lapply(out, "[", 3))

if (length(pair_order) > 1){
  for (i in 1:length(pair_order)){
    colnames(results_objs[[i]]) <- paste(colnames(results_objs[[i]]), pair_order[i], sep = ".")
    results_objs[[i]]$Geneid <- rownames(results_objs[[i]])
  }
  results_df <- Reduce(function(x, y) merge(x, y, by="Geneid"), results_objs)
  b <- addWorksheet(wb, "All")
  b <- writeData(wb, sheet="All", x=results_df, rowNames=T)
}

#tmp4 <- plyr::join_all(results_objs, by=0, type="full")



saveWorkbook(wb, file=paste0(analysis$config$analysis, "_DESeq2_results.xlsx"), overwrite=T)
#rm(wb)
# #for (a in 1:ncol(pairs)){
# #  p1[a] <- plot_ma(pairs[,a], analysis$ddsDrop)
# #}
# c1 <- apply(pairs, 2, plot_ma, analysis$ddsDrop)
# d1 <- ggarrange(plotlist=c1, common.legend=T)
# annotate_figure(d1, bottom="padj < 0.1, | FoldChange | > 1.3")
```


```{r}

analysis$rldDrop <- rlog(analysis$ddsDrop, blind = FALSE, fitType = "parametric")

analysis$assayRlogForGSEA <- assay(analysis$rldDrop)
# filter low/no-expression genes
analysis$assayRlogForGSEA <- analysis$assayRlogForGSEA[rowMeans(analysis$assayRlogForGSEA)>0,]
write_tsv(data.frame(Name = str_remove(rownames(analysis$assayRlogForGSEA), "[.].*"), Description = "na", analysis$assayRlogForGSEA), path = paste0("rlog_forGSEA_", analysis$config$analysis, ".txt"))
analysis$clsLinesGroup <- c(paste0(c(length(analysis$rldDrop$Group),length(unique(analysis$rldDrop$Group)),1), collapse = " "), paste0(c("#",unique(as.vector(analysis$rldDrop$Group))), collapse = " "), paste0(analysis$rldDrop$Group, collapse = " "))
write_lines(analysis$clsLinesGroup, path = paste0("Group_",analysis$config$analysis,".cls"))

```


Heatmap of all significant (Padj <`r analysis$config$alpha`) genes identified from all treatment comparisons. Expresion values are log transformed and normalized by the average expression in the control group.

```{r fig.height=22, fig.width=11}

# read in the DESeq2 regularized log expression for Seq1
rlogSeq1_PlotData <- read_tsv("rlog_forGSEA_p22074_Peijian_Analysis.txt") %>% select(-Description) %>% mutate(Name=make.names(Name, unique=T)) %>% column_to_rownames("Name") %>% as.matrix() 
colnames(rlogSeq1_PlotData) <- str_replace_all(colnames(rlogSeq1_PlotData),"\\.","-")
# read in the Sample information for Seq1

#all(match(colnames(rlogSeq1_PlotData), analysis$sampleTable$FileID))
# rlogSeq1_SampleData <- read_csv("rlogSeq1_SampleData_p20067_Ana.csv", col_types = "fffffd")
# all(match(colnames(rlogSeq1_PlotData),rlogSeq1_SampleData$SampleID)) # check match and order
# # read in the DESeq2 regularized log expression for Seq2
# rlogSeq2_PlotData <- read_tsv("rlogSeq2_forGSEA_p20067_Ana.txt") %>% select(-Description) %>% column_to_rownames("Name") %>% as.matrix()
# colnames(rlogSeq2_PlotData) <- str_replace_all(colnames(rlogSeq2_PlotData),"\\.","-")
# # read in the Sample information for Seq2
# rlogSeq2_SampleData <- read_csv("rlogSeq2_SampleData_p20067_Ana.csv", col_types = "fffffd")
# all(match(colnames(rlogSeq2_PlotData),rlogSeq2_SampleData$SampleID)) # check match and order
hmap <- rlogSeq1_PlotData[rownames(rlogSeq1_PlotData) %in% sig.genes,]
baseline_means <- rowMeans2(hmap[, analysis$rldDrop@colData$Group %in% analysis$config$designBaseline])
hmap <- hmap - baseline_means
Heatmap(hmap, show_row_names = F, heatmap_legend_param = list(title="RLE"))#[,!(analysis$rldDrop@colData$Group %in% analysis$config$designBaseline)])


```

# Individual gene analysis

No genes specified yet

```{r}
if (!is.null(analysis$config$geneList[[1]])){
  plot_genelist <- function(gene){
    plotCountsGKT(analysis$ddsDrop,gene, intgroup = c("Group")) + geom_sina() + scale_y_log10()
  }
  lapply(analysis$config$geneList, plot_genelist)
}
#plotCounts(analysis$ddsDrop, analysis$config$geneList, intgroup="Group")
```


## GSEA

Waiting on confirmation of experimental design

