---
title: "bulk_RNAseq_QC"
author: "ENPRC Gencore"
date:  "`r Sys.Date()`"
output: 
  rmdformats::robobook:
    fig_width: 9
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE,
                      message = FALSE, 
                      echo=FALSE,
                      fig.width = 6,
                      fig.height = 4,
                      dev='png')
library(tidyverse)    ## General logic and processing
library(yaml)         ## Parse config
library(DESeq2)       ## Makes us have a job
library(tools)
#library(circlize)    ## For colorRamp2 in heatmaps
library(openxlsx)     ## Writing count table outputs
library(here)         ## For consistent directory structures
##library(plotly)     ## If you want interactive PCA
library(ggrepel)      ## for PCA plot repel text
library(plyr)


##remotes::install_github('yerkes-gencore/gencore-bulk',force = TRUE)
library(gencoreBulk)  ## See the git repo for instructions on using this


## sometimes they conflict with other loaded packages
counts <- DESeq2::counts
here <- here::here
```



```{r reset_analysis_obj, eval=TRUE, include=FALSE}
# This is moved out of the setup block to avoid accidentally overwriting your
# analysis object when restarting a session
analysis = list()
```

# Load config

```{r load_config, include=FALSE}
# Store config with results object for record keeping
analysis$qc_config <- yaml.load_file(here("config/QC_config.yml"))
```

# Load samplesheet

```{r sample_metadata_setup}
## See the helper_script generate_experimental_design_sheet.R to help generate a table,
## or modify the template in the config folder and save it as 'exp_design'
analysis$sampleTable <- read.table(here("config/exp_design.txt"),
                                   header=T, 
                                   stringsAsFactors = TRUE)

## Possible types of processing to arrange or factorize terms
#analysis$sampleTable <- analysis$sampleTable %>% dplyr::arrange(across(all_of(analysis$qc_config$sampleGrouping)))
# analysis$sampleTable <- analysis$sampleTable %>%
#   mutate(across(.), as.factor)

sample_metadata_table <- knitr::kable((analysis$sampleTable), position='center', table.attr="style='width:100%;'", format='html')# %>%
    #kable_styling(full_width = TRUE)
# Store the order you'd like samples to appear in figure
# Order is determined by sampleGrouping parameters in config
# But you could also manually specify it here if there aren't easy rules
# sample_order <- fct_inorder(as.character(analysis$sampleTable$SampleID))

## for compatibility with filepaths
analysis$sampleTable$FileID <- as.character(analysis$sampleTable$FileID)

sample_metadata_table
```

# Load counts tables

```{r read_counts_tables, include=FALSE}
STARreadSuffix <- "_ReadsPerGene.out.tab"
readfiles <- sapply(
      analysis$sampleTable$FileID,
      function(sid) {
       paste0(dir(analysis$qc_config$alignmentDir,
                  pattern = sid, full.names = TRUE),
              "/", sid, STARreadSuffix)})
          
outs <- parseReadPerGeneFiles(readfiles, analysis$qc_config$libraryType)

mapBins <- outs$map_bins
raw_counts <- outs$read_counts

# order counts as they are in the sampletable
raw_counts <- raw_counts[,analysis$sampleTable$FileID]
```

# Load GTF

```{r gene_names, include=TRUE}
geneInfoTab <- read_tsv(file.path(analysis$qc_config$referenceDir,
                                  "geneInfo.tab"),
                        skip = 1,
                        col_names = c("gene_id","gene_symbol","gene_type"))

## Remove samples if there are fewer total fragment counts than features
## could catch a technical error if data isn't loading correctly
## (but you should see this in the mapping plot)
lowcounts <- colSums(raw_counts) < nrow(geneInfoTab)
stopifnot(!sum(lowcounts)>0)

gene_names <- plyr::mapvalues(rownames(raw_counts), 
                        from=geneInfoTab$gene_id,
                        to=geneInfoTab$gene_symbol)

## save this in case you want to write it out
raw_counts_ensembl <- raw_counts
rownames(raw_counts) <- make.names(gene_names, unique=TRUE)
```

# Create DESeq object

```{r create_dds, include=FALSE}
## safety check to make sure metadata and samples are properly associated
stopifnot(colnames(raw_counts) == analysis$sampleTable$FileID)

analysis$dds <- DESeqDataSetFromMatrix(raw_counts,
                                       analysis$sampleTable,
                                       design = as.formula( ~ 1))
```

# Mapping bins plots
```{r mapping_plot, fig.width=8, fig.height=6}
analysis$plots$mapping_bins_plot <- mappingBinsPlot(mapBins)
analysis$plots$mapping_bins_plot
```

# Remove samples

```{r remove_samples}
## Removing samples specified in the config. 
## This occurs after the mapBin plot just to have record of the 
## removed sample and possibly why it was removed, 
## but it could be adjusted to be upstream.

if (!is.null(unlist(analysis$qc_config$dropSamples))){
  mapBins <- mapBins[
    ,!colnames(mapBins) %in% analysis$qc_config$dropSamples]
  analysis$dds <- analysis$dds[
    ,!(colnames(analysis$dds) %in% analysis$qc_config$dropSamples)]
  # sample_order <- sample_order[
  #   !(sample_order %in% as.factor(analysis$qc_config$dropSamples))]
  out <- paste0("Sample(s)\n",
                paste0(analysis$qc_config$dropSamples, sep=", ", collapse = ''),
                "\nbeing removed from downstream analysis per user specification.") 
    cat(out)
} 
```

# VST and RLD

```{r vst_and_ddsDrop}
analysis$dds <- DESeq(analysis$dds, parallel = TRUE)
assays(analysis$dds)$vst <- varianceStabilizingTransformation(analysis$dds,
                                                         blind = FALSE,
                                                         fitType = "parametric")

assays(analysis$dds)$rld <- rlog(analysis$dds,
                                 blind = FALSE,
                                 fitType = "parametric")
```

# RLE plots

Library size normalization of read counts with hypothesis that most genes are 
not differentially expressed

```{r rle, fig.width = 6, fig.height=5}
rle_plots <- checkRLE(analysis$dds)
analysis$plots$RLE_raw <- rle_plots$RLE_raw
analysis$plots$RLE_norm <- rle_plots$RLE_norm
```

# PCA

Exploring projections of expression profiles onto low dimensional space using a 
limited number of highly variable features. This will give us an estimate of 
similarity between samples, but will not account for all genes.

```{r pca, fig.width=12, fig.height=8, warning=FALSE}
analysis$plots$pca <- plotPCAFromConfig(analysis)
print(analysis$plots$pca)
```

# Write tables

```{r}
writeCountTables(analysis,
                 normalized = FALSE,
                 generate_GSEA_app_tables = FALSE,
                 write_sample_table = TRUE)
```

# Save object

```{r}
if (!dir.exists(here('rds'))){dir.create(here('rds'))}
saveRDS(analysis, here('rds/analysis_post_QC.Rds'))
```

# Render report

```{r}
rmarkdown::render(here::here('scripts/bulk_rnaseq_qc_template.format.Rmd'),
                  output_file = 'QC_report.html',
                  output_dir = here::here('reports'),
                  params = list(
                    title = 'Bulk RNA QC Report')
                  )
```

