---
title: "`r project_name`: Report `r report_number` - QC"
author: |
  | `r report_author`
  | ENPRC Genomics Core
date: "`r Sys.Date()`"
output: 
  rmdformats::robobook:
    fig_width: 9
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=TRUE, message=TRUE,
                      fig.width = 8, fig.height = 6,
                      cache=FALSE, cache.lazy = FALSE,
                      fig.align = 'center')#, include=FALSE)
library(kableExtra)
```

```{=html}
<style type="text/css">
.book .book-body .page-inner {
  max-width: 1600px;
  margin-left: auto;
  margin-right: auto;
}
details > summary {
  display: list-item;
}
</style>
```

In this report, ``r report_filename_QC``, we walk through the results of running a quality control analysis on the provided data.

## 1 - Study design overview {.tabset}

<!-- Here's a template for presenting two different alternative designs requested by the client. In many cases, only one is necessary, so just delete Design B. -->

Note that there are three tables below, which you can navigate by clicking the tabs.

### Sample table

```{r, message=FALSE, warning=FALSE}
bulk$md$sampleTable %>%
  select(SampleID, SubjectID, grp, grp_desc, day, response) %>%
  knitr::kable() %>% kableExtra::kable_styling()
```

### Design A

```{r, message=FALSE, warning=FALSE}
bulk$md$sampleTable %>%
  select(SampleID, SubjectID, grp, grp_desc, day) %>%
  group_by(grp, grp_desc, day) %>%
    summarize(SubjectIDs = paste0(SubjectID, collapse = ", "),
              n_subjects = n()) %>%
  knitr::kable() %>% kableExtra::kable_styling()
```

### Design B

```{r, message=FALSE, warning=FALSE}
bulk$md$sampleTable %>%
  filter(grp == "grp2") %>%
  select(SampleID, SubjectID, grp, grp_desc, day, response) %>%
  group_by(grp, grp_desc, day, response) %>%
    summarize(SubjectIDs = paste0(SubjectID, collapse = ", "),
              n_subjects = n()) %>%
  knitr::kable() %>% kableExtra::kable_styling()
```


## 2 - Mapping stats

Outcomes from mapping reads to reference genome. Reads need to be 'identified' or successfully mapped to a single feature to be used for gene expression analysis.

```{r mapping_plot, fig.width=10, fig.height=6}
mappingBinsPlot(bulk$md$mapBins) +
  theme(axis.text.x = element_text(hjust = 1))
```

## 3 - Gene filtering by expression

Genes with low or zero expression across most samples are not informative and removing them improves False Discovery Rate control and differential expression model fit. Here are the results of filtering genes based on expression. Genes retained for downstream analysis are in green (non-zero expression) or yellow (zero expression), and genes filtered from analysis are in orange (non-zero expression) and red (zero expression). 

```{r, fig.width=10, fig.height=10}
plotFilterByExpr(bulk_prefilt$dge, keep.exprs)
```


<!-- ## Relative log expression (RLE) normalization -->
<!-- [TDB] -->


## 4 - PCA

Exploring projections of expression profiles onto low dimensional space using a limited number of highly variable genes. This will give us an estimate of similarity between samples, but will not account for all genes.

```{r, fig.width=8, fig.height=8}
ggplotMDS(bulk$dge, group = "grp.day", gene.selection = "common",
          sampleID = "sampleID", color = "grp", path = "SubjectID")
```
