---
title: "01-limmaEdgeR-QC-runfile"
author: "ENPRC Genomics Core"
date:  "`r Sys.Date()`"
output: 
  rmdformats::robobook:
    fig_width: 9
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = TRUE,
                      message = TRUE, 
                      echo=FALSE, 
                      fig.width = 8, 
                      fig.height = 4)
```

# Libraries

```{r}
library(tidyverse)
library(limma)
library(edgeR)
library(gencoreBulk)
library(here)
# library(msigdbr)
source(here("R/voomByGroup.R"))
```


# Outline of workflow

1. Read in data and create `bulk` object
  - Read in counts from STAR output
  - Add sample metadata to `bulk` object
  - Add design to `bulk` object
2. Filter by expression
  - Run `FilterByExpr()`
  - Visualize library sizes and gene counts before and after filtering
3. PCA
  - Using ggplot version of output of `plotMDS()`

## Report metadata

These variables help assign filenames to outputs and reports so you can rerun this script with different parameters without mixing up or overwriting files you want to track or compare.

```{r}
project_name <- "p23131_Tomas_RNAseq"
report_number <- "01"
report_title_root <- "QC"
```


# 0 - Custom functions

To do: Move this to `R/` and/or `gencoreBulk` if these prove useful

```{r}
getVoomByGroupData <- function(bulk, group, ...) {
  vbg <- voomByGroup(counts = bulk$dge$counts, 
                     design = bulk$md$design, 
                     group = bulk$dge$samples[[group]], 
                     plot = "combine", save.plot = TRUE)
  
  vbg_plot_data <- 
    lapply(names(vbg$voom.line), function(grp) {
      as_tibble(vbg$voom.line[[grp]]) %>%
        mutate(group = grp) %>%
        distinct()
    }) %>% bind_rows()
    
  return(vbg_plot_data)
}

plotVoomByGroupData <- function(vbg_data, ...) {
  vbg_data %>%
    ggplot(data = ., aes(x = x, y = y, color = group)) +
    geom_line() +
    scale_color_brewer(palette = "Paired") +
    xlab("log2( count size + 0.5 )") +
    ylab("Sqrt( standard deviation )") +
    theme_classic()
}

ggplotMDS <- function(dge, group = "group", sampleID = "sampleID", gene.selection = "common", ...) {
  mds_data <- plotMDS(dge, top = 500,
                      labels = dge$samples[[group]], 
                      col = col.group, plot = FALSE, 
                      gene.selection, ...)
  
  mds_xy <- mds_data[c("x","y")] %>% as_tibble()
  mds_xy[[sampleID]] <- colnames(dge)
  # mds_xy[[sampleID]] <- str_extract(colnames(dge), "(^[A-Za-z0-9]*)")
  mds_xy <- mds_xy %>% full_join(dge$samples, by = sampleID)
  
  x_varex <- round(mds_data$var.explained[1]*100, digits = 0)
  y_varex <- round(mds_data$var.explained[2]*100, digits = 0)
  
  mds_xy %>%
  ggplot(aes(x = .data$x, y = .data$y, color = .data[[group]], label = .data[[sampleID]])) +
    geom_text() + 
    scale_color_brewer(palette = "Paired") +
    xlab(paste0(mds_data$axislabel, " 1 (", x_varex, "%)")) +
    ylab(paste0(mds_data$axislabel, " 2 (", y_varex, "%)")) +
    theme_classic() +
    theme(aspect.ratio = 1) +
    coord_cartesian(clip = "off") +
    ggtitle(ifelse(gene.selection == "common", "PCA", "MDS"))
}
```


# 1 - Read in data and create `bulk` object

```{r}
## `bulk` object will contain project metadata, sample metadata, counts tables, and model fits
bulk <- list()
```

## 1.1 - Read/set project config

May move this to a yaml as for the DESeq2 version. Leaving here for clarity.

```{r}
bulk$md$project <- "p23131_Tomas_RNAseq"
bulk$md$refDir <- "/yerkes-cifs/runs/Genome_references/composite/macaca_mulatta_siv/Mmul10_SIVsm804ECL757/STAR-2.7.9a"
bulk$md$alignDir <- "/yerkes-cifs/runs/Analysis/2023_Analyses/p23131_Tomas/processing/STAR-2.7.9a"
bulk$md$dropSamples <- c()
bulk$md$libraryType <- "antisense"
bulk$md$STARreadSuffix <- "_ReadsPerGene.out.tab"

## Subset of samples from p23131_Tomas project for demonstration purposes
bulk$md$sampleTable <- tibble(
  fileID = c("p23131-s008_RJu20-April-6-2023-Day-0-Inclacumab",
               "p23131-s029_RLn21-May-18-2023-Day-0-Inclacumab",
               "p23131-s006_RRk19-April-6-2023-Day-0-Inclacumab",
               "p23131-s018_RJu20-April-27-2023-Day-21-post-infection-Inclacumab",
               "p23131-s039_RLn21-June-15-2023-Day-21-post-infection-Inclacumab",
               "p23131-s016_RRk19-April-27-2023-Day-21-post-infection-Inclacumab",
               "p23131-s009_RCn20-April-6-2023-Day-0",
               "p23131-s030_RDd21-May-18-2023-Day-0",
               "p23131-s026_RHy19-May-18-2023-Day-0",
               "p23131-s019_RCn20-April-27-2023-Day-21-post-infection",
               "p23131-s040_RDd21-June-15-2023-Day-21-post-infection",
               "p23131-s036_RHy19-June-15-2023-Day-21-post-infection"),
  animalID = c("RJu20", "RLn21", "RRk19", 
               "RJu20", "RLn21", "RRk19",
               "RCn20", "RDd21", "RHy19",
               "RCn20", "RDd21", "RHy19"),
  day = c("D0", "D0", "D0", 
          "D21", "D21", "D21",
          "D0", "D0", "D0", 
          "D21", "D21", "D21"),
  trt = c("I", "I", "I", 
          "I", "I", "I",
          "U", "U", "U", 
          "U", "U", "U")
)

bulk$md$sampleTable <- bulk$md$sampleTable %>%
  mutate(grp = paste0(trt, "_", day),
         sampleID = paste0(animalID, "_", grp))

```

## 1.2 - Read in counts from STAR output and create DGEList object

### Read in counts and mapping stats
```{r}
## Assemble full path of each read file
bulk$md$sampleTable$readPath <- 
  here(bulk$md$alignDir, bulk$md$sampleTable$fileID, paste0(bulk$md$sampleTable$fileID, bulk$md$STARreadSuffix))

## STAR puts mapping stats in the top 4 lines of the counts table files *ReadsPerGene.out.tab, so can't use edgeR::readDGE() out of the box
counts_tmp <- parseReadPerGeneFiles(
  file.paths = setNames(bulk$md$sampleTable$readPath, 
                        nm = bulk$md$sampleTable$sampleID), 
  library.type = bulk$md$libraryType)

## Save mapping stats into project metadata for plotting later
bulk$md$mapBins <- counts_tmp$map_bins
```

### Create DGEList
```{r}
bulk$dge <- DGEList(counts = counts_tmp$read_counts, 
                    samples = bulk$md$sampleTable %>% dplyr::select(-readPath), # remove paths from dge sample metadata because they are really long
                    group = bulk$md$sampleTable$grp)

## Don't need this anymore
rm(counts_tmp)
```

### Assign gene symbols from reference gtf as gene IDs
```{r}
## STAR output may use Ensembl IDs as gene IDs, in which case, use gtf to assign gene symbols where possible
geneInfoTab <- read_tsv(here(bulk$md$refDir, "geneInfo.tab"),
                        skip = 1,
                        col_names = c("gene_id","gene_symbol","gene_type"),
                        show_col_types = FALSE)

# Remove samples if there are fewer total fragment counts than features
lowcounts <- colSums(bulk$dge$counts) < nrow(geneInfoTab)
stopifnot(!sum(lowcounts)>0)

gene_names <- plyr::mapvalues(rownames(bulk$dge$counts), 
                              from = geneInfoTab$gene_id,
                              to = geneInfoTab$gene_symbol)

# Save original ensembl names in case you want to write it out
raw_counts_ensembl <- bulk$dge$counts
rownames(bulk$dge$counts) <- make.names(gene_names, unique=TRUE)
```

## 1.3 - Add design to `bulk` object

See `https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html` for detailed guidance on how to choose a design matrix for a given experiment.
Here we use combined two factors, `trt` and `day`, into a single factor `grp`. Note that this is a repeated measures experiment and we will fit the model using a random effect downstream.
```{r}
bulk$md$design <- model.matrix(~ 0 + grp, data = bulk$md$sampleTable)
```

## 1.4 - Correct gene names based on gtf

```{r gene_names, include=TRUE}
geneInfoTab <- read_tsv(here(bulk$md$refDir, "geneInfo.tab"),
                        skip = 1,
                        col_names = c("gene_id","gene_symbol","gene_type"),
                        show_col_types = FALSE)

## Remove samples if there are fewer total fragment counts than features
## Could catch a technical error if data isn't loading correctly (but you should see this in the mapping plot)
lowcounts <- colSums(bulk$dge$counts) < nrow(geneInfoTab)
stopifnot(!sum(lowcounts)>0)

gene_names <- plyr::mapvalues(rownames(bulk$dge$counts), 
                              from=geneInfoTab$gene_id,
                              to=geneInfoTab$gene_symbol)

## Save original ensembl names in case you want to write it out
raw_counts_ensembl <- bulk$dge$counts
rownames(bulk$dge$counts) <- make.names(gene_names, unique=TRUE)
```

## 1.5 - Plot mapping stats

```{r mapping_plot, fig.width=8, fig.height=6}
mappingBinsPlot(bulk$md$mapBins)
```

# 2 - Filter by expression

```{r}
keep.exprs <- filterByExpr(bulk$dge)
```

## 2.1 - Gene filtering by expression
```{r}
bulk$dge$counts %>%
  as_tibble(rownames = "gene") %>%
  left_join(tibble(gene = names(keep.exprs), keep = keep.exprs), 
            by = c("gene")) %>%
  pivot_longer(cols = -c(gene, keep), names_to = "sampleID", values_to = "counts") %>%
  mutate(is_zero = counts == 0) %>%
  group_by(sampleID, keep) %>%
  summarize(keep_zero = sum(is_zero & keep),
            keep_nonzero = sum(!is_zero & keep),
            remove_zero = sum(is_zero & !keep),
            remove_nonzero = sum(!is_zero & !keep)) %>%
  pivot_longer(cols = c(starts_with("keep_"), starts_with("remove_")), names_to = "cat", values_to = "n_genes") %>%
  mutate(cat = factor(cat, levels = rev(c("keep_nonzero", "keep_zero", "remove_nonzero", "remove_zero")))) %>%
  ggplot(aes(y = sampleID, x = n_genes, fill = cat)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("keep_nonzero"="green2", "keep_zero"="yellow2",
                               "remove_nonzero"="orange", "remove_zero"="red"),
                    breaks = c("keep_nonzero", "keep_zero",
                               "remove_nonzero", "remove_zero")) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

## 2.2 - Library sizes of discarded and kept data
```{r, fig.height=8, fig.width=8}
bulk$dge$counts %>%
  as_tibble(rownames = "gene") %>%
  left_join(tibble(gene = names(keep.exprs), keep = keep.exprs), 
            by = c("gene")) %>%
  pivot_longer(cols = -c(gene, keep), names_to = "sampleID", values_to = "counts") %>%
  group_by(sampleID, keep) %>%
    summarize(lib.size = sum(counts)) %>%
  ungroup() %>%
  ggplot(aes(y = sampleID, x = lib.size, fill = keep)) +
  geom_bar(stat="identity") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

bulk$dge$counts %>%
  as_tibble(rownames = "gene") %>%
  left_join(tibble(gene = names(keep.exprs), keep = keep.exprs), 
            by = c("gene")) %>%
  pivot_longer(cols = -c(gene, keep), names_to = "sampleID", values_to = "counts") %>%
  group_by(sampleID, keep) %>%
    summarize(lib.size = sum(counts)) %>%
  ungroup() %>%
  ggplot(aes(y = sampleID, x = lib.size, fill = keep)) +
  geom_bar(stat="identity") +
  scale_x_log10() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```

## 2.3 - Apply the filter
```{r}
## Save the prefiltered data in case you want to compare
bulk_prefilt <- bulk

## Apply the filter
bulk$dge <- bulk$dge[keep.exprs, , keep.lib.sizes=FALSE]
```


# 3 - Variance-mean trends and PCA
  - Using ggplot version of output of `plotMDS()`

## 3.1 - Mean-variance trend by group (i.e. voomByGroup)

```{r}
vbg_plot <- getVoomByGroupData(bulk, group = "grp") %>%
    plotVoomByGroupData()
```

## 3.2 - PCA plot

**To do: add options for elipses, lines, shapes and color scales as in `gencoreBulk::plotPCAFromConfig()`**

```{r}
ggplotMDS(bulk$dge, group = "grp", gene.selection = "common")
```

## 3.3 - MDS plot

Multidimensional scaling plot; distances correspond to BCV between pairs of samples. Pairwise BCVs were computed from the 500 most heterogeneous genes. See ?limma::plotMDS() for details.

[McCarthy et al. 2012](https://academic.oup.com/nar/article/40/10/4288/2411520)
> Biological coefficient of variation is the coefficient of variation with which the (unknown) true abundance of the gene varies between replicate RNA samples. It represents the CV that would remain between biological replicates if sequencing depth could be increased indefinitely. 

```{r}
ggplotMDS(bulk$dge, group = "grp", gene.selection = "pairwise")
```

# 4 - Export count table and write bulk object rds

## Create a subdir name and keep it consistent across dirs

```{r}
## Subdir name for `outputs` and `rds`
subdir <- paste0("report-", report_number, "-", report_title_root)
## Formatted date for filenames
getDate <- function() {
  format(Sys.Date(), "%Y%m%d")
}
```

## Export count table

```{r}
if (!dir.exists(here("outputs/", subdir))) {
  dir.create(here("outputs/", subdir))
}

bulk$dge$counts %>% 
  as_tibble(rownames = "gene") %>%
  write_csv(., here("outputs/", subdir, paste0("counts-", getDate(), ".rds")))
```

## Write `bulk` rds
```{r}
if (!dir.exists(here("rds/", subdir))) {
  dir.create(here("rds/", subdir))
}

bulk_prefilt %>%
  write_rds(., here("rds/", subdir, paste0("bulk_prefilt-", getDate(), ".rds")))
bulk %>%
  write_rds(., here("rds/", subdir, paste0("bulk-", getDate(), ".rds")))
```

