---
title: "01-limmaEdgeR-QC+DE+GSEA-runfile"
author: "ENPRC Genomics Core"
date:  "`r Sys.Date()`"
output: 
  rmdformats::robobook:
    fig_width: 9
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = TRUE,
                      message = TRUE, 
                      echo=FALSE, 
                      fig.width = 8, 
                      fig.height = 4)
```

# Libraries

```{r}
library(tidyverse)
library(limma)
library(edgeR)
library(gencoreBulk)
library(readr)
library(here)
library(qvalue)
library(msigdbr)
library(fgsea)
source(here("R/limmaEdgeR_functions.R")) # will move these to gencoreBulk package
source(here("R/heatmap_functions.R"))
```


# Explanation of report workflow

This runfile is where we can run all the code required for the analysis, which will populate our environment with the objects used in the format file. We typically run as much of the more heavier computation in the runfile as possible so that it doesn't take long to render the format file, since we may need to render iteratively as we adjust the format, fix typos, write explanatory notes, or add report sections.

This version of the runfile uses edgeR and limma as the main DE workhorses, and fgsea as the functional analysis tool. The script is organized into three parts running from (1) Quality Control (data input, filtering by expression, PCA plotting and exporting counts table) to (2) Differential Expression (defining design formula and contrasts, running diagnostics and generating DEG plots and summaries) and (3) Gene Set Enrichment Analysis (against various different gene set lists). In part 4 we render the report.

You can see examples of rendered reports at `/yerkes-cifs/runs/analyst/micah/Analysis/2023_Analyses/p23181_Abbie/p23181_Abbie_Analysis/reports`.

If you haven't discussed the study design and goals with the clients much yet, you can simply comment out parts 2 and 3 in the runfile and format file and render only the QC stage. Then you can discuss the QC with them in the first meeting and talk in more detail about their contrasts of interest, any expected batch effects or other nuisance variables, etc. and come back here, uncomment parts 2 and 3 and discuss the updated report for the next meeting.

The `## Report metadata` section controls the name of the report. When iterating reports, you could start with `report_number <- "01"` for the initial report and `report_number <- "02"` for the full QC + DE + GSEA report, for example. If you are simply fixing typos or adding a minor set of plots using the same base dataset, you could keep the report number the same and rely on the report timestamp in the filename to communicate the version. If the project includes several independent datasets that progress through the same steps (e.g. report "01-QC", "02-QC+DE+GSEA"), you could add a label to the report number to reflect this while preserving the progression of the analysis in the number (e.g. "thai_01-QC", "thai_02-QC+DE+GSEA", "atl_01-QC", "atl_02-QC+DE+GSEA"). This is at least what worked well for the `p23181_Abbie`, on which this template is based.

Note that the example dataset here is a cross-sectional study with several contrasts of interest, but this script can be adjusted to better handle projects with a single contrast. Just uncomment the code chunk section 4.2.

## Report metadata

These variables help assign filenames to outputs and reports so you can rerun this script with different parameters without mixing up or overwriting files you want to track or compare.

```{r}
project_name <- "p24000_Template"
report_number <- "cross-01"
report_author <- "Micah Fletcher"
```

## Custom functions

We may want to move these to the gencoreBulk package if they are stable and versitile enough, but they can stay here if adjustments need to be made on the fly or if they are still in development.

```{r}
## Part 1:
ggplotMDS <- function(dge, sampleID = "sampleID", gene.selection = "common", 
                          dims = c(1,2),
                          color = NULL, shape = NULL, size = 4,
                          ellipse = NULL, path = NULL,
                          show.labels = TRUE, label.size = 4, custom.labels = NULL, 
                          alpha = 1, ...) {
  # Get mds data from edgeR::plotMDS()
  mds_data <- limma::plotMDS(dge, top = 500, plot = FALSE, 
                             gene.selection = gene.selection, dim.plot = dims, ...)
  
  mds_xy <- mds_data[c("x","y")]
  mds_xy[[sampleID]] <- colnames(dge)
  # mds_xy[[sampleID]] <- str_extract(colnames(dge), "(^[A-Za-z0-9]*)")
  mds_xy <- dplyr::as_tibble(mds_xy) %>% dplyr::full_join(dge$samples, by = sampleID)
  
  x_varex <- round(mds_data$var.explained[dims[1]]*100, digits = 0)
  y_varex <- round(mds_data$var.explained[dims[2]]*100, digits = 0)
  
  if (!is.null(custom.labels)) {
    mds_xy <- mds_xy %>%
      mutate(sampleID = ifelse(sampleID %in% custom.labels, sampleID, NA))
  }
  
  mds_xy %>%
    ggplot(aes(x = .data$x, y = .data$y)) +
    geom_point(
      aes(
        color = (
          if (!is.null(color)) { .data[[color]] } else { NULL }
          ),
        shape = (
          if (!is.null(shape)) { .data[[shape]] } else { NULL }
          ),
        # text = (if (is.null(analysis$qc_config$pcaMapping$hover)) {
        #   NULL
        # } else {
        #   .data[[analysis$qc_config$pcaMapping$hover]]
        # })
      ),
      size = size,
      alpha = alpha
    ) +
    (if (show.labels) { ggrepel::geom_text_repel(aes(label = .data[["sampleID"]]), box.padding = 0.5, na.rm = TRUE) } else { NULL }) +
    (if (!is.null(path)) { geom_path(aes(linetype = .data[[path]])) } else { NULL }) +
    (if (!is.null(ellipse)) { stat_ellipse(aes(color = .data[[ellipse]]), type = "norm", level = 0.67)} else { NULL }) +
    labs(color = color, shape = shape) +
    xlab(paste0(mds_data$axislabel, " ", dims[1], " (", x_varex, "%)")) +
    ylab(paste0(mds_data$axislabel, " ", dims[2]," (", y_varex, "%)")) +
    theme_classic() +
    theme(aspect.ratio = 1) +
    coord_cartesian(clip = "off") +
    ggtitle(ifelse(gene.selection == "common", "PCA", "MDS"))
}

## Part 2:
fitVoomLm <- function(bulkObj, design = bulk$md$design, contr.matrix, 
                      block = NULL, sample.weights = TRUE, 
                      var.design = NULL, var.group = NULL, plotVoom = TRUE,
                      ebayes_fun = "eBayes", fc = 1.2) {
    bulkObj$fit <- edgeR::voomLmFit(counts = bulkObj$dge, # Defaults to the normalized (effective) library sizes in counts if counts is a DGEList or to the columnwise count totals if counts is a matrix.
                                    design = design,
                                    block = block,
                                    sample.weights = sample.weights,
                                    var.design = var.design,
                                    var.group = var.group,
                                    plot = plotVoom)
  bulkObj$fit.contr <- limma::contrasts.fit(bulkObj$fit, contrasts = contr.matrix)
  if (ebayes_fun == "eBayes") {
    bulkObj$fit.contr <- limma::eBayes(bulkObj$fit.contr, robust=TRUE)
  } else if (ebayes_fun == "treat") {
    bulkObj$fit.contr <- limma::treat(bulkObj$fit.contr, fc = fc, robust=TRUE)
  }
  limma::plotSA(bulkObj$fit.contr, main="Final model: Mean-variance trend")
  return(bulkObj)
}

createResTable <- function (fit, contr.matrix, ebayes_fun = "eBayes") {
    resultsTables_list <- list()
    for (contrast in colnames(fit$coefficients)) {
      if (ebayes_fun == "eBayes") {
        resultsTables_list[[contrast]] <- limma::topTable(fit, 
            coef = contrast, n = Inf) %>% dplyr::rename(log2FoldChange = .data$logFC, 
            pvalue = .data$P.Value, padj = .data$adj.P.Val)
      } else if (ebayes_fun == "treat") {
        resultsTables_list[[contrast]] <- limma::topTreat(fit, 
            coef = contrast, n = Inf) %>% dplyr::rename(log2FoldChange = .data$logFC, 
            pvalue = .data$P.Value, padj = .data$adj.P.Val)
      }
    }
    resultsTable <- lapply(resultsTables_list, function(one_tbl) {
        one_tbl %>% tibble::rownames_to_column(var = "gene")
    }) %>% dplyr::bind_rows(.id = "contrast") %>% dplyr::as_tibble() %>% 
        dplyr::mutate(contrast = forcats::fct(contrast, levels = colnames(contr.matrix)))
    return(resultsTable)
}
```


# Part 1 - QC

## 1 - Read in data and create `bulk` object

```{r}
## `bulk` object will contain project metadata, sample metadata, counts tables, and model fits
bulk <- list()
```

### 1.1 - Read/set project config

May move this to a yaml as for the DESeq2 version. Leaving here for clarity.

```{r}
bulk$md$project <- project_name
bulk$md$refDir <- "/yerkes-cifs/runs/Genome_references/homo_sapiens/GRCh38/ensembl_107/star_index_2.7.9a/"
bulk$md$alignDir <- "/yerkes-cifs/runs/analyst/micah/Analysis/2023_Analyses/p23181_Abbie/p23181_Abbie_Processing/S2.7.9a_GRCh38-107/"
bulk$md$dropSamples <- c()
bulk$md$libraryType <- "unstranded"
bulk$md$STARreadSuffix <- "_ReadsPerGene.out.tab"

manifest_tbl <- read_csv("/yerkes-cifs/runs/analyst/micah/Analysis/2023_Analyses/p23181_Abbie/p23181_Abbie_Analysis/notes/03-25-24 Kelley TGS Metadata.csv", show_col_types = FALSE, col_types = "cccccc")

orig_colnames <- colnames(manifest_tbl)
colnames(manifest_tbl) <- c("sampleID", "batch", "cohort", "site", "visit", "gender")
manifest_tbl <- manifest_tbl %>%
  mutate(batch = case_match(batch, "3" ~ NA, .default = batch),
         cohort = case_match(cohort, "1" ~ "cross", "2" ~ "long", "3" ~ NA),
         site = case_match(site, "1" ~ "atl", "2" ~ "thai", "3" ~ NA),
         visit = case_match(visit, "0" ~ "cohort1", "1" ~ "pre", "2" ~ "post", "3" ~ NA),
         gender = case_match(gender, "0" ~ "cohort2", "1" ~ "CGM", "2" ~ "TGW", "3" ~ NA),
         sampleID = str_replace_all(sampleID, " ", "-")) %>%
  filter(!is.na(batch)) # remove samples that were sequenced twice

sampleTable_cross <- tibble(fileID=dir(bulk$md$alignDir)) %>%
  mutate(sampleID = str_sub(fileID,start = 13),
         subjectID = str_extract(fileID,"(ATL|BKK)-[0-9]-[0-9]*")) %>%
  left_join(manifest_tbl, by = "sampleID") %>%
  mutate(gender_site = paste0(gender, "_", site)) %>%
  mutate(sampleID = ifelse(fileID == "p24069-s001_ATL-1-002", "ATL-1-002-p24069", sampleID),
         subjectID = ifelse(fileID == "p24069-s001_ATL-1-002", "ATL-1-002-p24069", subjectID)) %>% # duplicate sample, Greg wants to check how close they are and then remove
  filter(cohort == "cross") %>%
  filter(sampleID != "ATL-1-002-p24069") # removing from report as both samples look fine

sampleTable_longit <- tibble(fileID=dir(bulk$md$alignDir)) %>%
  mutate(sampleID = str_sub(fileID,start = 13),
         subjectID = str_extract(fileID,"(ATL|BKK)-[0-9]-[0-9]*")) %>%
  left_join(manifest_tbl, by = "sampleID") %>%
  filter(cohort == "long")
```


```{r}
bulk$md$sampleTable <- sampleTable_cross

## Rearrange metadata for visualizations (cross-sectional)
bulk$md$sampleTable <- bulk$md$sampleTable %>%
  arrange(site, gender) %>%
  mutate(batch = fct(batch, levels = c("1", "2")),
         gender_site = fct(gender_site) %>% fct_inorder(),
         sampleID = fct(sampleID) %>% fct_inorder(),
         subjectID = fct(subjectID) %>% fct_inorder()) %>%
  select(fileID, sampleID, subjectID, batch, gender_site)

# ## Rearrange metadata for visualizations (longitudinal)
# bulk$md$sampleTable <- bulk$md$sampleTable %>%
#   arrange(site, visit) %>%
#   mutate(visit = fct(visit, levels = c("pre", "post")),
#          sampleID = fct(sampleID) %>% fct_inorder(),
#          subjectID = fct(subjectID) %>% fct_inorder()) %>%
#   select(fileID, sampleID, subjectID, visit)
```

### 1.2 - Read in counts from STAR output and create DGEList object

#### Read in counts and mapping stats
```{r}
## Assemble full path of each read file
bulk$md$sampleTable$readPath <- 
  here(bulk$md$alignDir, bulk$md$sampleTable$fileID, paste0(bulk$md$sampleTable$fileID, bulk$md$STARreadSuffix))

## STAR puts mapping stats in the top 4 lines of the counts table files *ReadsPerGene.out.tab, so can't use edgeR::readDGE() out of the box
counts_tmp <- parseReadPerGeneFiles(
  file.paths = setNames(bulk$md$sampleTable$readPath, 
                        nm = bulk$md$sampleTable$sampleID), 
  library.type = bulk$md$libraryType)

## Save mapping stats into project metadata for plotting later
bulk$md$mapBins <- counts_tmp$map_bins
```

#### Create DGEList
```{r}
bulk$dge <- DGEList(counts = counts_tmp$read_counts, 
                    samples = bulk$md$sampleTable %>% dplyr::select(-readPath), # remove paths from dge sample metadata because they are really long
                    group = bulk$md$sampleTable$gender_site)

bulk$dge$samples$lib.size <- colSums(bulk$dge$counts)
bulk$dge <- normLibSizes(bulk$dge)
bulk$dge$cpm <- cpm(bulk$dge$counts)
bulk$dge$lcpm <- cpm(bulk$dge$counts, log = TRUE)

## Don't need this anymore
rm(counts_tmp)
```

#### Assign gene symbols from reference gtf as gene IDs
```{r}
## STAR output may use Ensembl IDs as gene IDs, in which case, use gtf to assign gene symbols where possible
geneInfoTab <- read_tsv(here(bulk$md$refDir, "geneInfo.tab"),
                        skip = 1,
                        col_names = c("gene_id","gene_symbol","gene_type"),
                        show_col_types = FALSE)

# Remove samples if there are fewer total fragment counts than features
lowcounts <- colSums(bulk$dge$counts) < nrow(geneInfoTab)
stopifnot(!sum(lowcounts)>0)

gene_names <- plyr::mapvalues(rownames(bulk$dge$counts), 
                              from = geneInfoTab$gene_id,
                              to = geneInfoTab$gene_symbol)

# Save original ensembl names in case you want to write it out
raw_counts_ensembl <- bulk$dge$counts
rownames(bulk$dge$counts) <- make.names(gene_names, unique=TRUE)
```

### 1.3 - Add design to `bulk` object

See `https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html` for detailed guidance on how to choose a design matrix for a given experiment.
```{r}
bulk$md$design <- stats::model.matrix(~ 0 + gender_site, data = bulk$md$sampleTable)
```

### 1.4 - Plot mapping stats

```{r mapping_plot, fig.width=16, fig.height=6}
mappingBinsPlot(bulk$md$mapBins) +
  theme(axis.text.x = element_text(hjust = 1, size = 6))
```

## 2 - Filter by expression

"The filtering should be based on the grouping factors or treatment factors that will be involved in the differential expression tests tested for, rather than on blocking variables that are not of scientific interest in themselves." 

So you may need to change the `design = bulk$md$design` to `group = {factor of biological interest}` if you've included nuisance variables in the design. See page 14 of EdgeR manual for details: https://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf

```{r}
keep.exprs <- filterByExpr(bulk$dge, design = stats::model.matrix(~ 0 + gender_site, data = bulk$md$sampleTable))
```

### 2.1 - Gene filtering by expression
```{r, fig.height = 6, fig.width = 16}
sampleID_colors <- tibble(
  sampleID = levels(bulk$dge$samples$sampleID)) %>%
  mutate(color = "black")

bulk$dge$counts %>%
    dplyr::as_tibble(rownames = "gene") %>%
    dplyr::left_join(dplyr::tibble(gene = names(keep.exprs), keep = keep.exprs), 
                     by = c("gene")) %>%
    tidyr::pivot_longer(cols = -c("gene", "keep"), names_to = "sampleID", values_to = "counts") %>%
    dplyr::mutate(sampleID = forcats::fct(.data$sampleID, levels = levels(bulk$dge$samples$sampleID)),
                  is_zero = .data$counts == 0) %>%
    dplyr::group_by(.data$sampleID) %>%
    dplyr::summarize(keep_zero = sum(.data$is_zero & .data$keep),
                     keep_nonzero = sum(!.data$is_zero & .data$keep),
                     remove_zero = sum(.data$is_zero & !.data$keep),
                     remove_nonzero = sum(!.data$is_zero & !.data$keep)) %>%
    dplyr::ungroup() %>%
    tidyr::pivot_longer(cols = c(starts_with("keep_"), starts_with("remove_")), names_to = "cat", values_to = "n_genes") %>%
    dplyr::mutate(cat = forcats::fct(.data$cat, levels = rev(c("keep_nonzero", "keep_zero", "remove_nonzero", "remove_zero")))) %>%
    ggplot(aes(x = .data$sampleID, y = .data$n_genes, fill = .data$cat)) +
    geom_bar(stat="identity") +
    scale_fill_manual(name = "Filtering",
                      values = c("keep_nonzero"="green2", "keep_zero"="yellow2",
                                 "remove_nonzero"="orange", "remove_zero"="red"),
                      breaks = c("keep_nonzero", "keep_zero",
                                 "remove_nonzero", "remove_zero"),
                      labels = c("Kept non-zero", "Kept zero",
                                 "Removed non-zero", "Removed zero")) +
    ylab("Gene counts") + xlab("Sample ID") +
    theme_bw() +
    theme(axis.text.x = ggtext::element_markdown(angle = 90, vjust = 0.5, hjust = 1, size=6, color = sampleID_colors$color))
```


### 2.2 - Library sizes of discarded and kept data
```{r, fig.height=6, fig.width=16}
bulk$dge$counts %>%
  as_tibble(rownames = "gene") %>%
  left_join(tibble(gene = names(keep.exprs), keep = keep.exprs), 
            by = c("gene")) %>%
  pivot_longer(cols = -c(gene, keep), names_to = "sampleID", values_to = "counts") %>%
  mutate(sampleID = fct(sampleID, levels = levels(bulk$dge$samples$sampleID))) %>%
  group_by(sampleID, keep) %>%
    summarize(lib.size = sum(counts)) %>%
  ungroup() %>%
  ggplot(aes(x = sampleID, y = lib.size, fill = keep)) +
  geom_bar(stat="identity") +
  ylab("Library size (sum of counts)") + xlab("Sample ID") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=6))

bulk$dge$counts %>%
  as_tibble(rownames = "gene") %>%
  left_join(tibble(gene = names(keep.exprs), keep = keep.exprs), 
            by = c("gene")) %>%
  pivot_longer(cols = -c(gene, keep), names_to = "sampleID", values_to = "counts") %>%
  mutate(sampleID = fct(sampleID, levels = levels(bulk$dge$samples$sampleID))) %>%
  group_by(sampleID, keep) %>%
    summarize(lib.size = sum(counts)) %>%
  ungroup() %>%
  ggplot(aes(x = sampleID, y = lib.size, fill = keep)) +
  geom_bar(stat="identity") +
  scale_y_log10() +
  ylab("log10(Library size)") + xlab("Sample ID") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=6))
```


### 2.3 - Apply the filter
```{r}
## Save the prefiltered data in case you want to compare
bulk_prefilt <- bulk

## Apply the filter
bulk$dge <- bulk$dge[keep.exprs, , keep.lib.sizes=FALSE]

## Recalculate library size, norm factors, cpm and lcpm
bulk$dge$samples$lib.size <- colSums(bulk$dge$counts)
bulk$dge <- normLibSizes(bulk$dge)
bulk$dge$cpm <- cpm(bulk$dge$counts)
bulk$dge$lcpm <- cpm(bulk$dge$counts, log = TRUE)
```


## 3 - PCA

### 3.1 - PCA plot

```{r, fig.width=8, fig.height=8}
dims_list <- list("1v2"=c(1,2), "2v3"=c(2,3), "3v4"=c(3,4))

## Generate PCA plots
pcas <- lapply(dims_list, function(dims) {
  ggplotMDS(bulk$dge, dims = dims, group = "gender_site", gene.selection = "common", show.labels = FALSE,
          sampleID = "sampleID", color = "gender_site", shape = "batch", alpha = 0.50)
})

pcas

## Find genes associated with each of the top 4 dimensions
pca12_data <- limma::plotMDS(bulk$dge, top = 500, plot = FALSE, 
                             gene.selection = "common", dim.plot = c(1,2))

pca12_design <- cbind(Intercept=1,xaxis=pca12_data$x,yaxis=pca12_data$y)
pca12_fit <- edgeR::voomLmFit(bulk$dge, design = pca12_design)
pca12_fit <- limma::eBayes(pca12_fit)
pc1_degs <- limma::topTable(pca12_fit, coef="xaxis", number = 30)
pc2_degs <- limma::topTable(pca12_fit, coef="yaxis", number = 30)

pca34_data <- limma::plotMDS(bulk$dge, top = 500, plot = FALSE, 
                             gene.selection = "common", dim.plot = c(1,2))

pca34_design <- cbind(Intercept=1,xaxis=pca34_data$x,yaxis=pca34_data$y)
pca34_fit <- edgeR::voomLmFit(bulk$dge, design = pca34_design)
pca34_fit <- limma::eBayes(pca34_fit)
pc3_degs <- limma::topTable(pca34_fit, coef="xaxis", number = 30)
pc4_degs <- limma::topTable(pca34_fit, coef="yaxis", number = 30)
```


### 3.2 - MDS plot

Multidimensional scaling plot; distances correspond to BCV between pairs of samples. Pairwise BCVs were computed from the 500 most heterogeneous genes. See ?limma::plotMDS() for details.

[McCarthy et al. 2012](https://academic.oup.com/nar/article/40/10/4288/2411520)
> Biological coefficient of variation is the coefficient of variation with which the (unknown) true abundance of the gene varies between replicate RNA samples. It represents the CV that would remain between biological replicates if sequencing depth could be increased indefinitely.

```{r, fig.width=8, fig.height=8}
## Generate plots
mdss <- lapply(dims_list, function(dims) {
  ggplotMDS(bulk$dge, dims = dims, group = "gender_site", gene.selection = "pairwise", show.labels = FALSE,
          sampleID = "sampleID", color = "gender_site", shape = "batch", alpha = 0.50)
})

mdss

## Find genes associated with each of the top 4 dimensions
mds12_data <- limma::plotMDS(bulk$dge, top = 500, plot = FALSE, 
                             gene.selection = "pairwise", dim.plot = c(1,2))

mds12_design <- cbind(Intercept=1,xaxis=mds12_data$x,yaxis=mds12_data$y)
mds12_fit <- edgeR::voomLmFit(bulk$dge, design = mds12_design)
mds12_fit <- limma::eBayes(mds12_fit)
mdsdim1_degs <- limma::topTable(mds12_fit, coef="xaxis", number = 30)
mdsdim2_degs <- limma::topTable(mds12_fit, coef="yaxis", number = 30)

mds34_data <- limma::plotMDS(bulk$dge, top = 500, plot = FALSE, 
                             gene.selection = "pairwise", dim.plot = c(1,2))

mds34_design <- cbind(Intercept=1,xaxis=mds34_data$x,yaxis=mds34_data$y)
mds34_fit <- edgeR::voomLmFit(bulk$dge, design = mds34_design)
mds34_fit <- limma::eBayes(mds34_fit)
mdsdim3_degs <- limma::topTable(mds34_fit, coef="xaxis", number = 30)
mdsdim4_degs <- limma::topTable(mds34_fit, coef="yaxis", number = 30)
```

## 4 - Batch-corrected PCAs

### 4.1 - ComBat-seq

Use ComBat-seq from sva package to remove batch effects from counts for vizualizations
```{r}
bulk_CBs <- bulk
bulk_CBs$dge$counts <- sva::ComBat_seq(bulk$dge$counts, 
                                         batch = bulk$dge$samples$batch,
                                         group = bulk$dge$samples$gender_site)

## Recalculate library size, norm factors, cpm and lcpm
bulk_CBs$dge$samples$lib.size <- colSums(bulk_CBs$dge$counts)
bulk_CBs$dge <- normLibSizes(bulk_CBs$dge)
bulk_CBs$dge$cpm <- cpm(bulk_CBs$dge$counts)
bulk_CBs$dge$lcpm <- cpm(bulk_CBs$dge$counts, log = TRUE)
```

### 4.2 - PCA plot

```{r, fig.width=8, fig.height=8}
## Generate plots
pcas_CBs <- lapply(dims_list, function(dims) {
  ggplotMDS(bulk_CBs$dge, dims = dims, group = "gender_site", gene.selection = "common", show.labels = FALSE,
          sampleID = "sampleID", color = "gender_site", shape = "batch", alpha = 0.50)
})

pcas_CBs

## Find genes associated with each of the top 4 dimensions
pca12_data_CBs <- limma::plotMDS(bulk_CBs$dge, top = 500, plot = FALSE, 
                             gene.selection = "common", dim.plot = c(1,2))

pca12_design_CBs <- cbind(Intercept=1,xaxis=pca12_data_CBs$x,yaxis=pca12_data_CBs$y)
pca12_fit_CBs <- edgeR::voomLmFit(bulk_CBs$dge, design = pca12_design_CBs)
pca12_fit_CBs <- limma::eBayes(pca12_fit_CBs)
pc1_degs_CBs <- limma::topTable(pca12_fit_CBs, coef="xaxis", number = 30)
pc2_degs_CBs <- limma::topTable(pca12_fit_CBs, coef="yaxis", number = 30)

pca34_data_CBs <- limma::plotMDS(bulk_CBs$dge, top = 500, plot = FALSE, 
                             gene.selection = "common", dim.plot = c(1,2))

pca34_design_CBs <- cbind(Intercept=1,xaxis=pca34_data_CBs$x,yaxis=pca34_data_CBs$y)
pca34_fit_CBs <- edgeR::voomLmFit(bulk_CBs$dge, design = pca34_design_CBs)
pca34_fit_CBs <- limma::eBayes(pca34_fit_CBs)
pc3_degs_CBs <- limma::topTable(pca34_fit_CBs, coef="xaxis", number = 30)
pc4_degs_CBs <- limma::topTable(pca34_fit_CBs, coef="yaxis", number = 30)
```

### 4.3 - MDS plot

Multidimensional scaling plot; distances correspond to BCV between pairs of samples. Pairwise BCVs were computed from the 500 most heterogeneous genes. See ?limma::plotMDS() for details.

[McCarthy et al. 2012](https://academic.oup.com/nar/article/40/10/4288/2411520)
> Biological coefficient of variation is the coefficient of variation with which the (unknown) true abundance of the gene varies between replicate RNA samples. It represents the CV that would remain between biological replicates if sequencing depth could be increased indefinitely. 

```{r, fig.width=8, fig.height=8}
## Generate plots
mdss_CBs <- lapply(dims_list, function(dims) {
  ggplotMDS(bulk_CBs$dge, dims = dims, group = "gender_site", gene.selection = "pairwise", show.labels = FALSE,
          sampleID = "sampleID", color = "gender_site", shape = "batch", alpha = 0.50)
})

mdss_CBs

## Find genes associated with each of the top 4 dimensions
mds12_data_CBs <- limma::plotMDS(bulk_CBs$dge, top = 500, plot = FALSE, 
                             gene.selection = "pairwise", dim.plot = c(1,2))

mds12_design_CBs <- cbind(Intercept=1,xaxis=mds12_data_CBs$x,yaxis=mds12_data_CBs$y)
mds12_fit_CBs <- edgeR::voomLmFit(bulk_CBs$dge, design = mds12_design_CBs)
mds12_fit_CBs <- limma::eBayes(mds12_fit_CBs)
mdsdim1_degs_CBs <- limma::topTable(mds12_fit_CBs, coef="xaxis", number = 30)
mdsdim2_degs_CBs <- limma::topTable(mds12_fit_CBs, coef="yaxis", number = 30)

mds34_data_CBs <- limma::plotMDS(bulk_CBs$dge, top = 500, plot = FALSE, 
                             gene.selection = "pairwise", dim.plot = c(1,2))

mds34_design_CBs <- cbind(Intercept=1,xaxis=mds34_data_CBs$x,yaxis=mds34_data_CBs$y)
mds34_fit_CBs <- edgeR::voomLmFit(bulk_CBs$dge, design = mds34_design_CBs)
mds34_fit_CBs <- limma::eBayes(mds34_fit_CBs)
mdsdim3_degs_CBs <- limma::topTable(mds34_fit_CBs, coef="xaxis", number = 30)
mdsdim4_degs_CBs <- limma::topTable(mds34_fit_CBs, coef="yaxis", number = 30)
```

## 5 - Export count table and write bulk object rds

### 5.1 - Create a subdir name and keep it consistent across dirs

```{r}
## Subdir name for `outputs` and `rds`
subdir <- paste0("report-", report_number)
## Formatted date for filenames
getDate <- function() {
  format(Sys.Date(), "%Y%m%d")
}
```

### 5.2 - Export count table

```{r}
if (!dir.exists(here("outputs/", subdir))) {
  dir.create(here("outputs/", subdir))
}

bulk$dge$counts %>% 
  as_tibble(rownames = "gene") %>%
  write_csv(., here("outputs/", subdir, paste0("counts-", getDate(), ".rds")))
```

### 5.3 - Write `bulk` rds
```{r}
if (!dir.exists(here("rds/", subdir))) {
  dir.create(here("rds/", subdir))
}

bulk_prefilt %>%
  write_rds(., here("rds/", subdir, paste0("bulk_prefilt-", getDate(), ".rds")))
bulk %>%
  write_rds(., here("rds/", subdir, paste0("bulk-", getDate(), ".rds")))
bulk_CBs %>%
  write_rds(., here("rds/", subdir, paste0("bulk_CBs-", getDate(), ".rds")))

```

# Part 2 - DE

1. Define contrasts
  - Keep this separate from the `bulk` object for flexibility
2. DE diagnostics
  - Run `voomLmFit()` or `glmQLFit()`
  - Generate results tables
  - Plot MD
  - Plot p-value distributions
3. DE results
  - Summarize DEG counts in each contrast
  - Export results list to excel file

## 1 - Define FDR threshold

```{r}
fdr <- 0.05
```

## 2 - Define contrasts

See https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html for a detailed tutorial on how to set up design formulas, model matrices and contrast matrices for all of the commonly encountered RNAseq study designs.

<!-- Assuming we'll be using the `limma::duplicateCorrelation()` method of modeling individual as a random effect: -->
```{r}
bulk$md$design <- stats::model.matrix(~ 0 + gender_site, data = bulk$md$sampleTable)

colnames(bulk$md$design) <- gsub(x = colnames(bulk$md$design), "^gender_site", "")

contr.matrix <- makeContrasts(
  TGWvsCGM = ((TGW_atl - CGM_atl) + (TGW_thai - CGM_thai))/2,
  ATL_TGWvsCGM = TGW_atl - CGM_atl,
  THAI_TGWvsCGM = TGW_thai - CGM_thai,
  TGW_THAIvsATL = TGW_thai - TGW_atl,
  CGM_THAIvsATL = CGM_thai - CGM_atl,
  levels = colnames(bulk$md$design)
)
```

## 3 - DE diagnostics

### 3.1 - Run `voomLmFit()` or `glmQLFit()`

```{r}
bulk <- fitVoomLm(bulk,
                  design = bulk$md$design,
                  contr.matrix=contr.matrix,
                  sample.weights = TRUE)
```


### 3.2 - Generate results tables

```{r}
resultsTable <- createResTable(bulk$fit.contr, contr.matrix)
```

There are two easy ways to access tables split by contrast for different purposes using `for()` or `split()`
```{r}
# for (contr in unique(resultsTable$contrast)) {
#   resultsTable %>%
#     filter(contrast == contr)
# }
```

```{r}
# resultsTable %>%
#   split(., .$contrast)
```


### 3.3 - Plot MD

```{r, fig.width = 8, fig.height = 6}
md_data <- resultsTable %>%
    mutate(signif_dir = ifelse(padj >= fdr, "NotSig",
                           ifelse(log2FoldChange < 0, "Down", "Up")) %>%
             factor(., levels = c("NotSig", "Up", "Down")),
           signif = padj < fdr)
  
ggplot() +
  geom_point(data = filter(md_data, !signif),
             aes(x = AveExpr, y = log2FoldChange), 
             color = "black", size = 0.1) +
  geom_point(data = filter(md_data, signif), 
             aes(x = AveExpr, y = log2FoldChange, color = signif_dir), 
             size = 0.1) +
  ylim(-max(abs(md_data$log2FoldChange)), max(abs(md_data$log2FoldChange))) +
  scale_color_manual(values = c("black", "red", "blue"), drop = FALSE) +
  theme_bw() +
  facet_wrap(~contrast)
```


### 3.4 - Plot p-value distributions

```{r, fig.width = 8, fig.height = 6}
resultsTable %>%
      mutate(!!paste0("pBH < ", fdr) := ifelse(padj < fdr, "TRUE", "FALSE") %>%
               factor(., levels = c("TRUE", "FALSE"))) %>%
    ggplot(data = ., aes(x = pvalue, fill = .data[[paste0("pBH < ", fdr)]])) +
      geom_histogram(breaks = seq(0,1,0.01)) +
      scale_fill_manual(values = c("red", "black"), drop=FALSE) +
      theme_bw() +
    facet_wrap(~contrast, scales = "free_y")
```

### 3.5 - Calculate Storey q-value and plot diagnostics

#### Q value diagnostics
```{r}
resultsTable %>%
  split(., f = resultsTable$contrast) %>%
  lapply(., function(x) {
    qvalue(x$pvalue) %>% summary()
    qvalue(x$pvalue) %>% hist()
    qvalue(x$pvalue) %>% plot()
  })
```

#### Add q-value to results table

```{r}
resultsTable <- resultsTable %>%
  split(., f = resultsTable$contrast) %>%
  lapply(., function(res_contrast) {
    res_contrast$qvalue <-  qvalue(res_contrast$pvalue)$qvalues
    return(res_contrast)
  }) %>%
  bind_rows()
```

#### Re-define FDR if desired based on qvalue diagnostics

```{r}
# fdr <- 0.05
```

## 4 - DE results

### 4.1 - Summarize DEG counts in each contrast

```{r}
n_deg_summary <- resultsTable %>% 
  group_by(contrast) %>%
    dplyr::rename(l2fc = log2FoldChange, pBH = padj) %>%
    summarize(signif_down = sum(pBH < fdr & l2fc < 0),
              nonsignif = sum(pBH >= fdr),
              signif_up = sum(pBH < fdr & l2fc > 0)) %>%
  mutate(signif = signif_up + signif_down)
```

#### Table
```{r}
n_deg_summary %>%
  mutate(signif_signs = paste0(signif_up, " ↑ / ", signif_down, " ↓")) %>%
  select(contrast, signif_signs) %>%
  knitr::kable()
```

#### Bar plots
```{r, fig.width = 12, fig.height = 8}
n_deg_summary %>%
  pivot_longer(cols = c(signif_down, signif_up), 
               names_to = "sign", 
               values_to = "n_signif") %>%
  mutate(n_signif = ifelse(sign == "signif_down",
                           -n_signif, n_signif)) %>%
  ggplot(aes(y = fct_rev(contrast), x = n_signif, fill = sign)) +
  geom_bar(stat="identity") +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  theme_bw() +
  ylab("contrast")
```

### 4.2 - Export results list to excel file

### Create a subdir name and keep it consistent across dirs

```{r}
## Subdir name for `outputs` and `rds`
subdir <- paste0("report-", report_number)
## Formatted date for filenames
getDate <- function() {
  format(Sys.Date(), "%Y%m%d")
}
```

Create output dir for this report if it doesn't already exist
```{r}
if (!dir.exists(here("outputs/", subdir))) {
  dir.create(here("outputs/", subdir))
}
```

#### Export excel file with columns for padj and lfc for each contrast
```{r}
# When working with multiple contrasts, create a spreadsheet with a logfc and pval column for each contrast.
resultsTable_for_export <- unique(resultsTable$contrast) %>%
lapply(., function(contr) {
  resultsTable %>%
    filter(contrast == contr) %>%
    dplyr::select(gene, l2fc = log2FoldChange, pBH = padj, qvalue)%>%
    dplyr::rename_with(~ paste0(contr, "_", .x, recycle0 = TRUE),
                  all_of(c("l2fc", "pBH", "qvalue")))
}) %>%
  reduce(full_join, by = "gene")

# # We can do this more simply if we only have one contrast
# resultsTable_for_export <- resultsTable

resultsTable_filename <- paste0(
  project_name, "-",
  report_number, "-",
  "DE_results_tables-",
  getDate(), ".xlsx"
  )

c(list("Summary" = n_deg_summary), 
  list("ResultsTable" = resultsTable_for_export)) %>%
writexl::write_xlsx(here("outputs/", subdir, resultsTable_filename))
```

### 4.3 - Heatmap

This is a useful place to troubleshoot heatmaps for the format file, which will ultimately be generated by the harder to debug `knit_expand()` expressions.
```{r, fig.height = 12, fig.width = 24}
sample_tbl <- bulk$dge$samples %>%
  dplyr::select(sampleID, gender_site) %>%
  mutate(color = "black") %>%
  dplyr::rename(group = gender_site)
gene_tbl <- resultsTable %>%
  getTopDegTbl(., contrast_i = "TGWvsCGM", groupID = "group", 
                 arrange_by = "pvalue", direction = "unequal", padj_cutoff = NULL,
                 slice_n = 30, filter_pattern = NULL) %>%
    select(gene, padj, l2fc = log2FoldChange) %>%
    mutate(color = ifelse(is.na(padj), "grey",
                          ifelse(padj >= fdr, "black",
                                 ifelse(l2fc > 0, "red",
                                        ifelse(l2fc < 0, "blue",
                                               NA)))))
```


```{r, fig.height = 6, fig.width = 32}
print(
  plotHeatmap(exprs_mat = bulk$dge$lcpm, gene_tbl = gene_tbl, sample_tbl = sample_tbl,
              sampleID = "sampleID", cluster_genes = TRUE, cluster_samples = FALSE)
)
```

# Part 3 - GSEA

## 1 - GSEA

### 1.1 - Pre-rank DE stats

```{r}
prepGseaStats2 <- function(marker_table, p_col = "p_val", lfc_col = "avg_log2FC") {
  ranked_genes <- marker_table %>% filter(!is.na(.data[[p_col]]))
  ranked_genes$stat <- ranked_genes[[lfc_col]] * -log10(ranked_genes[[p_col]] + .Machine$double.xmin)
  gsea_stats <- setNames(ranked_genes$stat, nm = rownames(ranked_genes))
  gsea_stats <- sort(gsea_stats)
}

gsea_stats <- resultsTable %>%
  filter(!is.na(pvalue)) %>%
  mutate(rank_stat = log2FoldChange * -log10(pvalue + .Machine$double.xmin)) %>% 
  split(.$contrast) %>%
    lapply(., function(y) {
      y %>% select(gene, rank_stat) %>%
        deframe() %>%
        sort()
    })

```

### 1.2 - Prep gene lists

```{r}
msigH <- msigdbr(species = "Homo sapiens", category = "H") %>%
  dplyr::select(gene_symbol, gs_name) %>% unstack()

msigGOBP <- msigdbr(species = "Homo sapiens", category = "C5", subcategory = "GO:BP") %>%
  dplyr::select(gene_symbol, gs_name) %>% unstack()

msigCP <- c(
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:BIOCARTA") %>%
    dplyr::select(gene_symbol, gs_name) %>% unstack(),
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>%
    dplyr::select(gene_symbol, gs_name) %>% unstack(),
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:PID") %>%
    dplyr::select(gene_symbol, gs_name) %>% unstack(),
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>%
    dplyr::select(gene_symbol, gs_name) %>% unstack(),
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:WIKIPATHWAYS") %>%
    dplyr::select(gene_symbol, gs_name) %>% unstack()
)
```

### 1.3 - Run GSEA
```{r}
gsea_results <- list()

gsea_results$H <- gsea_stats %>%
  lapply(., function(stats_vec) {
    fgsea(pathways = msigH, stats = stats_vec,
          minSize = 15, maxSize = 100) %>% arrange(pval)
  })

gsea_results$GOBP <- gsea_stats %>%
  lapply(., function(stats_vec) {
    fgsea(pathways = msigGOBP, stats = stats_vec,
          minSize = 15, maxSize = 100) %>% arrange(pval)
  })

gsea_results$CP <- gsea_stats %>%
  lapply(., function(stats_vec) {
    fgsea(pathways = msigCP, stats = stats_vec,
          minSize = 15, maxSize = 100) %>% arrange(pval)
  })
```

### 1.4 - Export results table with all contrasts in a wide format as excel file
```{r}
gsea_results %>%
  lapply(., function(x) {
    bind_rows(x, .id = "contrast")
    }) %>%
  bind_rows(.id = "database")


## Format results for excel/reports
gsea_results_fmt <- lapply(names(gsea_results), function(db) {
    lapply(gsea_results[[db]], function(gsea_results_contrast) {
      gsea_results_contrast %>%
        as_tibble() %>%
        arrange(padj) %>%
        rowwise() %>%
        mutate(leading_edge = paste(leadingEdge, collapse = ", ")) %>%
        ungroup() %>%
        dplyr::select(pathway, ES, NES, log2err, pval, padj, leading_edge)
    })
})
names(gsea_results_fmt) <- names(gsea_results)
# Reorder contrast in order of levels to keep consistent order in reports
gsea_results_fmt <- gsea_results_fmt %>% 
  lapply(., function(results_db) {
    results_db <- results_db[levels(resultsTable$contrast)] 
})

# ## One sheet per database, each contrast side-by-side
# gseaResults_compact <- lapply(names(gsea_results), function(db) {
#     lapply(names(gsea_results[[db]]), function(contrast) {
#       gsea_results_fmt[[db]][[contrast]] %>%
#         dplyr::select(pathway, NES, pBH = padj) %>%
#         dplyr::rename_with(~ paste0(contrast, "_", .x, recycle0 = TRUE), all_of(c("NES", "pBH")))
#           }) %>%
#     purrr::reduce(full_join, by = "pathway")
# })
# names(gseaResults_compact) <- names(gsea_results_fmt)

GSEA_filename <- paste0(
  project_name, "-", 
  report_number, "-",
  "GSEA_results_tables", "-",
  getDate(), ".xlsx"
  )

gsea_results_fmt %>%
  lapply(., function(x) bind_rows(x, .id = "contrast")) %>%
  writexl::write_xlsx(here("outputs", subdir, GSEA_filename))
```

# Part 4 - Generate report

## 1 - Write rds

### Write `bulk` rds
```{r}
if (!dir.exists(here("rds/", subdir))) {
  dir.create(here("rds/", subdir))
}

bulk %>%
  write_rds(., here("rds/", subdir, paste0("bulk-", getDate(), ".rds")))
```


## 2 - Render report

### Set filenames

```{r}
report_filename <- paste0(project_name, "-", 
                          report_number, "-",
                          format(Sys.Date(), "%Y%m%d"), ".html")
```

### Number of genes in heatmaps

```{r}
n_genes <- 50
n_pathways <- 30
```


### Render
```{r}
rmarkdown::render(input = here("scripts/01-limmaEdgeR-QC+DE+GSEA-format.Rmd"),
                  output_file = report_filename,
                  output_dir = here('reports'))
```
